// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/room/mapping/HeightMapUpdateParser.ts
import { IMessageDataWrapper } from '../../../../../../core/communication/messages/IMessageDataWrapper';
import { IMessageParser } from '../../../../../../core/communication/messages/IMessageParser';

export function encodeTileValue(
    z: number,
    blocked: boolean,
    isRoomTile = true
): number {
    if (!isRoomTile || z < 0) return -1; // -1 = not a room tile
    let value = Math.round(z * 256); // same scaling as RoomHeightMapParser (0x0100)
    if (blocked) value |= 0x4000; // 0x4000 bit = stacking blocked
    return value;
}

export class HeightMapUpdateParser implements IMessageParser {
    private _width = 0;
    private _height = 0;
    private _cells: { x: number; y: number; z: number; state: number }[] = [];

    flush(): boolean {
        this._width = 0;
        this._height = 0;
        this._cells = [];
        return true;
    }

    parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        this._width = wrapper.readInt();
        this._height = wrapper.readInt();

        // 🔴 server sends this, so we must read it
        const count = wrapper.readInt();

        const cells: { x: number; y: number; z: number; state: number }[] = [];
        for (let i = 0; i < count; i++) {
            const x = wrapper.readShort();
            const y = wrapper.readShort();
            const z = wrapper.readByte(); // use readUnsignedByte() if your API has it
            const state = wrapper.readByte(); // ditto
            cells.push({ x, y, z, state });
        }

        this._cells = cells;

        // Optional debug to confirm it parses cleanly
        console.log(
            '[🧱 HeightMapUpdateParser] width, height, cells:',
            this._width,
            this._height,
            this._cells.length
        );

        return true;
    }

    get width() {
        return this._width;
    }
    get height() {
        return this._height;
    }
    get cells() {
        return this._cells;
    }
}
