// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/roleplay/BlackJackStateParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

/* ---- Types ---- */
export type BJState = 'INVITE' | 'PLAYING' | 'ENDED';

export interface BJMsg {
    text: string;
    level: 'info' | 'warn' | 'error';
}

export interface BlackjackStateDTO {
    state: BJState; // 1
    dealerReveal: boolean; // 2
    dealerCards: string[]; // 3
    dealerTotal: number; // 4
    playerHand: string[]; // 5
    playerTotal: number; // 6
    playerTurn: boolean; // 7
    canDouble: boolean; // 8
    canSplit: boolean; // 9
    insuranceOffered: boolean; // 10
    bet: number; // 11
    messages: BJMsg[]; // 12
    isDealer: boolean; // 13 NEW
}

export class BlackjackStateParser implements IMessageParser {
    private _data: BlackjackStateDTO | null = null;

    public flush(): boolean {
        this._data = null;
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        try {
            const state = wrapper.readString() as BJState; // 1
            const dealerReveal = wrapper.readBoolean(); // 2

            const dealerCount = wrapper.readInt(); // 3
            const dealerCards: string[] = [];
            for (let i = 0; i < dealerCount; i++)
                dealerCards.push(wrapper.readString());

            const dealerTotal = wrapper.readInt(); // 4

            const playerCount = wrapper.readInt(); // 5
            const playerHand: string[] = [];
            for (let i = 0; i < playerCount; i++)
                playerHand.push(wrapper.readString());

            const playerTotal = wrapper.readInt(); // 6
            const playerTurn = wrapper.readBoolean(); // 7
            const canDouble = wrapper.readBoolean(); // 8
            const canSplit = wrapper.readBoolean(); // 9
            const insuranceOffered = wrapper.readBoolean(); // 10
            const bet = wrapper.readInt(); // 11

            const msgCount = wrapper.readInt(); // 12
            const messages: BJMsg[] = [];
            for (let i = 0; i < msgCount; i++) {
                const text = wrapper.readString();
                const level = wrapper.readString() as 'info' | 'warn' | 'error';
                messages.push({ text, level });
            }

            const isDealer = wrapper.readBoolean(); // 13 NEW

            this._data = {
                state,
                dealerReveal,
                dealerCards,
                dealerTotal,
                playerHand,
                playerTotal,
                playerTurn,
                canDouble,
                canSplit,
                insuranceOffered,
                bet,
                messages,
                isDealer,
            };

            return true;
        } catch (err) {
            console.error('[BlackjackStateParser] failed:', err);
            this._data = null;
            return false;
        }
    }

    public get data(): BlackjackStateDTO | null {
        return this._data;
    }
}
