// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/roleplay/BlackJackStateParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

/* ---- Types ---- */
export type BJState = 'INVITE' | 'PLAYING' | 'ENDED';

export interface BJMsg {
    text: string;
    level: 'info' | 'warn' | 'error';
}

export interface BlackjackStateDTO {
    state: BJState; // 1) matches composer: "state"
    dealerReveal: boolean; // 2)
    dealerCards: string[]; // 3)
    dealerTotal: number; // 4)
    playerHand: string[]; // 5)
    playerTotal: number; // 6)
    playerTurn: boolean; // 7)
    canDouble: boolean; // 8)
    canSplit: boolean; // 9)
    insuranceOffered: boolean; // 10)
    bet: number; // 11)
    messages: BJMsg[]; // 12)
}

/* ---- Parser ---- */
export class BlackjackStateParser implements IMessageParser {
    private _data: BlackjackStateDTO | null = null;

    public flush(): boolean {
        this._data = null;
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        try {
            // 1) state
            const state = wrapper.readString() as BJState;

            // 2) dealerReveal
            const dealerReveal = wrapper.readBoolean();

            // 3) dealerCards
            const dealerCount = wrapper.readInt();
            const dealerCards: string[] = [];
            for (let i = 0; i < dealerCount; i++)
                dealerCards.push(wrapper.readString());

            // 4) dealerTotal
            const dealerTotal = wrapper.readInt();

            // 5) playerHand
            const playerCount = wrapper.readInt();
            const playerHand: string[] = [];
            for (let i = 0; i < playerCount; i++)
                playerHand.push(wrapper.readString());

            // 6) playerTotal
            const playerTotal = wrapper.readInt();

            // 7) playerTurn
            const playerTurn = wrapper.readBoolean();

            // 8) canDouble
            const canDouble = wrapper.readBoolean();

            // 9) canSplit
            const canSplit = wrapper.readBoolean();

            // 10) insuranceOffered
            const insuranceOffered = wrapper.readBoolean();

            // 11) bet
            const bet = wrapper.readInt();

            // 12) messages
            const msgCount = wrapper.readInt();
            const messages: BJMsg[] = [];
            for (let i = 0; i < msgCount; i++) {
                const text = wrapper.readString();
                const level = wrapper.readString() as 'info' | 'warn' | 'error';
                messages.push({ text, level });
            }

            this._data = {
                state,
                dealerReveal,
                dealerCards,
                dealerTotal,
                playerHand,
                playerTotal,
                playerTurn,
                canDouble,
                canSplit,
                insuranceOffered,
                bet,
                messages,
            };

            return true;
        } catch (err) {
            console.error('[BlackjackStateParser] failed:', err);
            this._data = null;
            return false;
        }
    }

    public get data(): BlackjackStateDTO | null {
        return this._data;
    }
}
