// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/roleplay/InventoryMessageParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';


export interface InventoryItem {
    id: number;
    item_id: number;
    user_id: number;
    slot: number;
    quantity: number;
    durability: number;
    max_durability: number; // <-- NEW
    icon_path: string;
    name: string;
    item_type:
        | 'weapon'
        | 'shield'
        | 'potion'
        | 'food'
        | 'drink'
        | 'tool'
        | 'crop';
    rarity: 'common' | 'uncommon' | 'rare' | 'epic' | 'legendary';
    stats: {
        damage: number;
        defense: number;
        healing: number;
    };
}

export class InventoryMessageParser implements IMessageParser {
    private _items: InventoryItem[] = [];

    public flush(): boolean {
        this._items = [];
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        const total = wrapper.readInt();

        const allowedTypes = [
            'weapon',
            'shield',
            'potion',
            'food',
            'drink',
            'tool',
            'crop', // <-- added to align with server
        ] as const;

        const allowedRarities = [
            'common',
            'uncommon',
            'rare',
            'epic',
            'legendary',
        ] as const;

        for (let i = 0; i < total; i++) {
            const id = wrapper.readInt(); // 1
            const item_id = wrapper.readInt(); // 2
            const user_id = wrapper.readInt(); // 3
            const slot = wrapper.readInt(); // 4
            const quantity = wrapper.readInt(); // 5

            const durability = wrapper.readInt(); // 6
            const max_durability = wrapper.readInt(); // 7 <-- NEW (must match server order)

            const iconPath = wrapper.readString(); // 8
            const name = wrapper.readString(); // 9

            const item_type_raw = wrapper.readString(); // 10
            const item_type = allowedTypes.includes(item_type_raw as any)
                ? (item_type_raw as InventoryItem['item_type'])
                : 'tool';

            const rarityRaw = wrapper.readString(); // 11
            const rarity = allowedRarities.includes(rarityRaw as any)
                ? (rarityRaw as InventoryItem['rarity'])
                : 'common';

            const damage = wrapper.readInt(); // 12
            const defense = wrapper.readInt(); // 13
            const healing = wrapper.readInt(); // 14

            this._items.push({
                id,
                item_id,
                user_id,
                slot,
                quantity,
                durability,
                max_durability,
                icon_path: iconPath,
                name,
                item_type,
                rarity,
                stats: {
                    damage,
                    defense,
                    healing,
                },
            });
        }

        return true;
    }

    public get items(): InventoryItem[] {
        return this._items;
    }
}
