// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/roleplay/ItemOfferPopupParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export class ItemOfferPopupParser implements IMessageParser {
    private _senderId = 0;
    private _senderName = '';
    private _itemName = '';
    private _itemIcon = '';
    private _itemCost = 0;

    public flush(): boolean {
        this._senderId = 0;
        this._senderName = '';
        this._itemName = '';
        this._itemIcon = '';
        this._itemCost = 0;
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        this._senderId = wrapper.readInt();
        this._senderName = wrapper.readString();
        this._itemName = wrapper.readString();
        this._itemIcon = wrapper.readString();
        this._itemCost = wrapper.readInt();
        return true;
    }

    public get senderId(): number {
        return this._senderId;
    }

    public get senderName(): string {
        return this._senderName;
    }

    public get itemName(): string {
        return this._itemName;
    }

    public get itemIcon(): string {
        return this._itemIcon;
    }
    public get itemCost(): number {
        return this._itemCost;
    }
}
