// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/roleplay/TaxiOpenListParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export interface TaxiDestinationData {
    roomId: number;
    virtualRoomId: number;
    virtualName: string; // virtual room name
    userCount: number; // number of users currently there
}

export class TaxiOpenListParser implements IMessageParser {
    private _currentRoomId = 0;
    private _currentVrid = 0;
    private _destinations: TaxiDestinationData[] = [];

    public flush(): boolean {
        this._destinations = [];
        this._currentRoomId = 0;
        this._currentVrid = 0;
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        // Header: player's current position
        this._currentRoomId = wrapper.readInt();
        this._currentVrid = wrapper.readInt();

        // Destination list
        const total = wrapper.readInt();
        this._destinations = new Array(total);

        for (let i = 0; i < total; i++) {
            const roomId = wrapper.readInt();
            const virtualRoomId = wrapper.readInt();
            const virtualName = wrapper.readString();
            const userCount = wrapper.readInt(); // new field for occupancy

            this._destinations[i] = {
                roomId,
                virtualRoomId,
                virtualName,
                userCount,
            };
        }

        return true;
    }

    /** Getter for parsed data */
    public get destinations(): TaxiDestinationData[] {
        return this._destinations;
    }

    /** Current user's room ID (for "You are here" marker) */
    public get currentRoomId(): number {
        return this._currentRoomId;
    }

    /** Current user's virtual room ID (for "You are here" marker) */
    public get currentVrid(): number {
        return this._currentVrid;
    }
}
