// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/roleplay/TaxiOpenListParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

/**
 * Data structure for a single taxi destination.
 * Matches TaxiDestination on the server.
 */
export interface TaxiDestinationData {
    roomId: number;
    virtualRoomId: number;
    displayName: string;
}

/**
 * Parser for TaxiOpenListComposer
 * Reads: [count] [roomId][virtualRoomId][displayName] * count
 */
export class TaxiOpenListParser implements IMessageParser {
    private _destinations: TaxiDestinationData[] = [];

    public flush(): boolean {
        this._destinations = [];
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        const total = wrapper.readInt();
        this._destinations = [];

        for (let i = 0; i < total; i++) {
            const roomId = wrapper.readInt();
            const virtualRoomId = wrapper.readInt();
            const displayName = wrapper.readString();

            this._destinations.push({
                roomId,
                virtualRoomId,
                displayName,
            });
        }

        return true;
    }

    /** Getter for parsed data */
    public get destinations(): TaxiDestinationData[] {
        return this._destinations;
    }
}
