// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/roleplay/TurfCaptureProgressParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export interface TurfCaptureProgressData {
    turfId: string; // "NORTH" | "SOUTH" | "EAST" | "WEST"
    virtualRoomId: number;
    active: boolean;
    percent: number; // 0..100
    contested: boolean;
    attackerGangId: number;
    attackerGangName: string;
    defenderGangId: number;
    defenderGangName: string;

    // Optional tail fields if your server adds them later (we ignore if missing)
    defenderPrimaryColor?: string; // "#RRGGBB"
    defenderSecondaryColor?: string; // "#RRGGBB"
}

export class TurfCaptureProgressParser implements IMessageParser {
    private _data: TurfCaptureProgressData = {
        turfId: '',
        virtualRoomId: 0,
        active: false,
        percent: 0,
        contested: false,
        attackerGangId: 0,
        attackerGangName: '',
        defenderGangId: 0,
        defenderGangName: '',
        defenderPrimaryColor: '',
        defenderSecondaryColor: '',
    };

    public flush(): boolean {
        this._data = {
            turfId: '',
            virtualRoomId: 0,
            active: false,
            percent: 0,
            contested: false,
            attackerGangId: 0,
            attackerGangName: '',
            defenderGangId: 0,
            defenderGangName: '',
            defenderPrimaryColor: '',
            defenderSecondaryColor: '',
        };
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        // MUST match server TurfCaptureProgressComposer order:
        // string turfId, int vRoomId, bool active, int percent, bool contested,
        // int attId, string attName, int defId, string defName
        this._data.turfId = wrapper.readString();
        this._data.virtualRoomId = wrapper.readInt();
        this._data.active = wrapper.readBoolean();
        this._data.percent = wrapper.readInt();
        this._data.contested = wrapper.readBoolean();
        this._data.attackerGangId = wrapper.readInt();
        this._data.attackerGangName = wrapper.readString();
        this._data.defenderGangId = wrapper.readInt();
        this._data.defenderGangName = wrapper.readString();

        // If your server has appended the 2 optional color strings,
        // they will be present; if not, reading them would throw.
        // So use a guarded read (Nitro wrapper has no "hasMore" -> try/catch).
        try {
            this._data.defenderPrimaryColor = wrapper.readString();
            this._data.defenderSecondaryColor = wrapper.readString();
        } catch {
            /* optional tail */
        }

        return true;
    }

    public get data(): TurfCaptureProgressData {
        return this._data;
    }
}
