// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/roleplay/UserStatsParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';
import { getPackedSettings } from 'http2';

export interface UserStatsData {
    health: number;
    maxHealth: number;
    energy: number;
    maxEnergy: number;
    hunger: number;
    maxHunger: number;
    aggression: number;
    xp: number;
    maxXP: number;
    strength: number;
    level: number;
    points: number;
    stamina: number;
    hunger_level: number;
    gathering: number;
    defense: number;
    punches_thrown: number;
    punches_landed: number;
    damage_inflicted: number;
    damage_received: number;
    shifts_worked: number;
    kills: number;
    deaths: number;
    arrests: number;
    isAggressive: boolean;
    working: boolean;
    gangName: string;
    dead: boolean;
}

export class UserStatsParser implements IMessageParser {
    private _stats: UserStatsData = {
        health: 0,
        maxHealth: 100,
        energy: 0,
        maxEnergy: 100,
        hunger: 0,
        maxHunger: 100,
        aggression: 0,
        xp: 0,
        maxXP: 100,
        strength: 0,
        level: 0,
        points: 0,
        stamina: 0,
        hunger_level: 0,
        gathering: 0,
        defense: 0,
        punches_thrown: 0,
        punches_landed: 0,
        damage_inflicted: 0,
        damage_received: 0,
        shifts_worked: 0,
        kills: 0,
        deaths: 0,
        arrests: 0,
        isAggressive: false,
        working: false,
        gangName: "",
        dead: false,
        
    };

    public flush(): boolean {
        this._stats = {
            health: 0,
            maxHealth: 100,
            energy: 0,
            maxEnergy: 100,
            hunger: 0,
            maxHunger: 100,
            aggression: 0,
            xp: 0,
            maxXP: 100,
            strength: 0,
            level: 0,
            points: 0,
            stamina: 0,
            hunger_level: 0,
            gathering: 0,
            defense: 0,
            punches_thrown: 0,
            punches_landed: 0,
            damage_inflicted: 0,
            damage_received: 0,
            shifts_worked: 0,
            kills: 0,
            deaths: 0,
            arrests: 0,
            isAggressive: false,
            working: false,
            gangName: "",
            dead: false,
        };
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        this._stats.health = wrapper.readInt();
        this._stats.maxHealth = wrapper.readInt();
        this._stats.energy = wrapper.readInt();
        this._stats.maxEnergy = wrapper.readInt();
        this._stats.hunger = wrapper.readInt();
        this._stats.maxHunger = wrapper.readInt();
        this._stats.aggression = wrapper.readInt();
        this._stats.xp = wrapper.readInt();
        this._stats.maxXP = wrapper.readInt();
        this._stats.strength = wrapper.readInt();
        this._stats.level = wrapper.readInt();
        this._stats.points = wrapper.readInt();
        this._stats.stamina = wrapper.readInt();
        this._stats.hunger_level = wrapper.readInt();
        this._stats.gathering = wrapper.readInt();
        this._stats.defense = wrapper.readInt();
        this._stats.punches_thrown = wrapper.readInt();
        this._stats.punches_landed = wrapper.readInt();
        this._stats.damage_inflicted = wrapper.readInt();
        this._stats.damage_received = wrapper.readInt();
        this._stats.shifts_worked = wrapper.readInt();
        this._stats.kills = wrapper.readInt();
        this._stats.deaths = wrapper.readInt();
        this._stats.arrests = wrapper.readInt();
        this._stats.isAggressive = wrapper.readBoolean();
        this._stats.working = wrapper.readBoolean();
        this._stats.gangName = wrapper.readString();
        this._stats.dead = wrapper.readBoolean();
        return true;
    }

    public get stats(): UserStatsData {
        return this._stats;
    }
}
