// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/roleplay/UserStatsParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export interface UserStatsData {
    userId: number;
    username: string;
    figure: string;

    health: number;
    maxHealth: number;
    energy: number;
    maxEnergy: number;
    hunger: number;
    maxHunger: number;
    aggressionMs: number;

    xp: number;
    maxXP: number;
    strength: number;
    level: number;
    points: number;
    stamina: number;
    hunger_level: number;
    gathering: number;
    defense: number;

    punches_thrown: number;
    punches_landed: number;
    damage_inflicted: number;
    damage_received: number;
    shifts_worked: number;
    kills: number;
    deaths: number;
    arrests: number;

    isAggressive: boolean;
    working: boolean;
    dead: boolean;
    gangName: string;

    cash: number;
    bank: number;
    nameIconId: number;

    healthlevel: number;

    jobTitle: string;
    corporationName: string;
    corporationIconUrl: string;
    weeklyShifts: number;
    totalShifts: number;
}

export class UserStatsParser implements IMessageParser {
    private _stats: UserStatsData = null;

    public flush(): boolean {
        this._stats = null;
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        const s: any = {};

        // === MATCH ORDER EXACTLY TO COMPOSER ===
        s.userId = wrapper.readInt();
        s.username = wrapper.readString();
        s.figure = wrapper.readString();

        s.health = wrapper.readInt();
        s.maxHealth = wrapper.readInt();
        s.energy = wrapper.readInt();
        s.maxEnergy = wrapper.readInt();
        s.hunger = wrapper.readInt();
        s.maxHunger = wrapper.readInt();
        s.aggressionMs = wrapper.readInt();

        s.xp = wrapper.readInt();
        s.maxXP = wrapper.readInt();
        s.strength = wrapper.readInt();
        s.level = wrapper.readInt();
        s.points = wrapper.readInt();
        s.stamina = wrapper.readInt();
        s.hunger_level = wrapper.readInt();
        s.gathering = wrapper.readInt();
        s.defense = wrapper.readInt();

        s.punches_thrown = wrapper.readInt();
        s.punches_landed = wrapper.readInt();
        s.damage_inflicted = wrapper.readInt();
        s.damage_received = wrapper.readInt();
        s.shifts_worked = wrapper.readInt();
        s.kills = wrapper.readInt();
        s.deaths = wrapper.readInt();
        s.arrests = wrapper.readInt();

        s.isAggressive = wrapper.readBoolean();
        s.working = wrapper.readBoolean();
        s.dead = wrapper.readBoolean();
        s.gangName = wrapper.readString();

        s.cash = wrapper.readInt();
        s.bank = wrapper.readInt();
        s.nameIconId = wrapper.readInt();

        s.healthlevel = wrapper.readInt();

        // --- NEW FIELDS ---
        s.jobTitle = wrapper.readString();
        s.corporationName = wrapper.readString();
        s.corporationIconUrl = wrapper.readString();
        s.weeklyShifts = wrapper.readInt();
        s.totalShifts = wrapper.readInt();

        this._stats = s;
        return true;
    }

    public get stats(): UserStatsData {
        return this._stats;
    }
}
