// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/roleplay/VirtualRoomInfoParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export interface VirtualRoomInfoData {
    virtualRoomId: number;
    name: string;
    type: string;
    credits: number;
    diamonds: number; // ✨ NEW
    inGameTime: string; // "HH:mm"
    phase: 'DAY' | 'DUSK' | 'NIGHT' | 'DAWN';
}

export class VirtualRoomInfoParser implements IMessageParser {
    private _data: VirtualRoomInfoData = {
        virtualRoomId: 0,
        name: '',
        type: 'default',
        credits: 0,
        diamonds: 0, // ✨ NEW
        inGameTime: '00:00',
        phase: 'DAY',
    };

    public flush(): boolean {
        this._data = {
            virtualRoomId: 0,
            name: '',
            type: 'default',
            credits: 0,
            diamonds: 0, // ✨ NEW
            inGameTime: '00:00',
            phase: 'DAY',
        };
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        // MUST match compose order:
        // id, name, type, credits, diamonds, inGameTime, phase
        this._data.virtualRoomId = wrapper.readInt();
        this._data.name = wrapper.readString();
        this._data.type = wrapper.readString();
        this._data.credits = wrapper.readInt();
        this._data.diamonds = wrapper.readInt(); // ✨ NEW
        this._data.inGameTime = wrapper.readString();
        this._data.phase = wrapper.readString() as VirtualRoomInfoData['phase'];

        return true;
    }

    public get data(): VirtualRoomInfoData {
        return this._data;
    }
}
