// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/roleplay/VirtualRoomInfoParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export interface VirtualRoomInfoData {
    virtualRoomId: number;
    name: string;
    type: string;
    credits: number;
    diamonds: number;
    inGameTime: string;
    phase: 'DAY' | 'DUSK' | 'NIGHT' | 'DAWN';
    userCount: number; // 👈 NEW
}

export class VirtualRoomInfoParser implements IMessageParser {
    private _data: VirtualRoomInfoData = {
        virtualRoomId: 0,
        name: '',
        type: 'default',
        credits: 0,
        diamonds: 0,
        inGameTime: '00:00',
        phase: 'DAY',
        userCount: 0,
    };

    public flush(): boolean {
        this._data = {
            virtualRoomId: 0,
            name: '',
            type: 'default',
            credits: 0,
            diamonds: 0,
            inGameTime: '00:00',
            phase: 'DAY',
            userCount: 0,
        };
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        // id, name, type, credits, diamonds, time, phase, onlineCount
        this._data.virtualRoomId = wrapper.readInt();
        this._data.name = wrapper.readString();
        this._data.type = wrapper.readString();
        this._data.credits = wrapper.readInt();
        this._data.diamonds = wrapper.readInt();
        this._data.inGameTime = wrapper.readString();
        this._data.phase = wrapper.readString() as VirtualRoomInfoData['phase'];
        this._data.userCount = wrapper.readInt(); // 👈 NEW

        return true;
    }

    public get data(): VirtualRoomInfoData {
        return this._data;
    }
}
