// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/roleplay/VirtualRoomInfoParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export interface VirtualRoomInfoData {
    virtualRoomId: number;
    name: string;
    type: string;
    credits: number;
}

export class VirtualRoomInfoParser implements IMessageParser {
    private _data: VirtualRoomInfoData = {
        virtualRoomId: 0,
        name: '',
        type: 'default',
        credits: 0,
    };

    public flush(): boolean {
        this._data = {
            virtualRoomId: 0,
            name: '',
            type: 'default',
            credits: 0,
        };
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        // must match server compose order exactly:
        // appendInt(id), appendString(name), appendString(type), appendInt(credits)
        this._data.virtualRoomId = wrapper.readInt();
        this._data.name = wrapper.readString();
        this._data.type = wrapper.readString();
        this._data.credits = wrapper.readInt();

        return true;
    }

    public get data(): VirtualRoomInfoData {
        return this._data;
    }
}
