// @nitrots/nitro-renderer/src/nitro/communication/NitroCommunicationManager.ts
import { IMessageEvent } from '../../core';
import { NitroManager } from '../../core/common/NitroManager';
import { IConnection } from '../../core/communication/connections/IConnection';
import { IConnectionStateListener } from '../../core/communication/connections/IConnectionStateListener';
import { SocketConnectionEvent } from '../../core/communication/events/SocketConnectionEvent';
import { ICommunicationManager } from '../../core/communication/ICommunicationManager';
import { IMessageConfiguration } from '../../core/communication/messages/IMessageConfiguration';
import { NitroEvent } from '../../core/events/NitroEvent';
import { Nitro } from '../Nitro';
import { NitroCommunicationDemo } from './demo/NitroCommunicationDemo';
import { NitroCommunicationDemoEvent } from './demo/NitroCommunicationDemoEvent';
import { INitroCommunicationManager } from './INitroCommunicationManager';
import { NitroMessages } from './NitroMessages';
import { InventoryMessageEvent } from '@nitrots/nitro-renderer';
import { InventoryStore } from '@nitrots/nitro-renderer/src/nitro/communication/messages/parser/roleplay/InventoryStore';
import { InventoryContext } from '../../../../../../src/api/contexts/inventory/InventoryContext';
import { UserStatsEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/UserStatsEvent';
import { ZeusHealthEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/ZeusHealthEvent';
import { updateInventory } from '../../../../../../src/api/managers/InventoryManager';
import { InspectUserStatsEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/InspectUserStatsEvent';
import { GangStatusEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/GangStatusEvent';
import { UserLiveFeedEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/UserLiveFeedEvent';
import { GatherProgressEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/GatherProgressEvent';
import { OpenSellModuleEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/OpenSellModuleEvent';
import { CancelGatherEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/CancelGatherEvent';
import { ItemOfferPopupEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/ItemOfferPopupEvent';
import { GangInviteEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/GangInviteEvent';
import { GangMembersEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/GangMembersEvent';
import { GangRanksEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/GangRanksEvent';
import { UpdateWorkStatusEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/UpdateWorkStatusEvent';
import { WantedListEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/WantedListEvent';
import { CorporationsListEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/CorporationsListEvent';
import { VirtualRoomInfoEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/VirtualRoomInfoEvent';
import { EquippedLoadoutEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/EquippedLoadoutEvent';
import { PaletteHexesEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/PaletteHexesEvent';
import { CorporationMembersEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/CorporationMembersEvent';
import { CorporationRanksEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/CorporationRanksEvent';
import { BlackjackModuleEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/BlackjackModuleEvent';
import { MarketplaceModuleEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/MarketplaceModuleEvent';
import { TimeOfDayEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/TimeOfDayEvent';
import { HighLowModuleEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/HighLowModuleEvent';
import { HighLowStateEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/HighLowStateEvent';
import { BlackjackStateEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/BlackjackStateEvent';
import { SlotMachineModuleEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/SlotMachineModuleEvent';
import { SlotMachineStateEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/SlotMachineStateEvent';
import { WheelStateEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/WheelStateEvent';
import { WheelModuleEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/WheelModuleEvent';
import { HotelAlertEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/HotelAlertEvent';
export class NitroCommunicationManager
    extends NitroManager
    implements INitroCommunicationManager, IConnectionStateListener
{
    private _communication: ICommunicationManager;
    private _connection: IConnection;
    private _messages: IMessageConfiguration;

    private _demo: NitroCommunicationDemo;

    constructor(communication: ICommunicationManager) {
        super();

        this._communication = communication;
        this._connection = null;
        this._messages = new NitroMessages();

        this._demo = new NitroCommunicationDemo(this);

        this.onConnectionOpenedEvent = this.onConnectionOpenedEvent.bind(this);
        this.onConnectionClosedEvent = this.onConnectionClosedEvent.bind(this);
        this.onConnectionErrorEvent = this.onConnectionErrorEvent.bind(this);
        this.onConnectionAuthenticatedEvent =
            this.onConnectionAuthenticatedEvent.bind(this);
    }

    protected onInit(): void {
        if (this._connection) return;
        console.log('OnIt from NitroCommunicationManager initialized!!');

        Nitro.instance.events.addEventListener(
            NitroCommunicationDemoEvent.CONNECTION_AUTHENTICATED,
            this.onConnectionAuthenticatedEvent
        );

        this._connection = this._communication.createConnection(this);

        this._connection.registerMessages(this._messages);

        this._connection.addEventListener(
            SocketConnectionEvent.CONNECTION_OPENED,
            this.onConnectionOpenedEvent
        );
        this._connection.addEventListener(
            SocketConnectionEvent.CONNECTION_CLOSED,
            this.onConnectionClosedEvent
        );
        this._connection.addEventListener(
            SocketConnectionEvent.CONNECTION_ERROR,
            this.onConnectionErrorEvent
        );

        if (this._demo) this._demo.init();

        this._connection.init(
            Nitro.instance.getConfiguration<string>('socket.url')
        );
    }

    protected onDispose(): void {
        if (this._demo) this._demo.dispose();

        if (this._connection) {
            this._connection.removeEventListener(
                SocketConnectionEvent.CONNECTION_OPENED,
                this.onConnectionOpenedEvent
            );
            this._connection.removeEventListener(
                SocketConnectionEvent.CONNECTION_CLOSED,
                this.onConnectionClosedEvent
            );
            this._connection.removeEventListener(
                SocketConnectionEvent.CONNECTION_ERROR,
                this.onConnectionErrorEvent
            );
        }

        Nitro.instance.events.removeEventListener(
            NitroCommunicationDemoEvent.CONNECTION_AUTHENTICATED,
            this.onConnectionAuthenticatedEvent
        );

        super.onDispose();
    }
    private onGatherProgressEvent(event: GatherProgressEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const duration = parser.duration;

        console.log('[📦] GatherProgressEvent → duration =', duration);

        // Dispatch to show the progress bar
        window.dispatchEvent(
            new CustomEvent('gather_progress', {
                detail: { duration },
            })
        );
    }
    private onEquippedLoadoutEvent(event: EquippedLoadoutEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const payload = {
            weapon: parser.weapon, // or null
            shield: parser.shield, // or null
            potion: parser.potion, // or null
        };

        // Dispatch to client UI
        window.dispatchEvent(
            new CustomEvent('equipped_loadout_update', { detail: payload })
        );
        console.log('🧰 EquippedLoadoutEvent ->', payload);
    }
    private onUserStatsEvent(event: UserStatsEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const stats = parser.stats;
        window.dispatchEvent(
            new CustomEvent('user_stats_stats', {
                detail: stats, // opponent stats
            })
        );
        //console.log('[📊] Received User Stats:', stats);

        // Dispatch to client UI — use your preferred method (event or context)
        window.dispatchEvent(
            new CustomEvent('user_stats_update', { detail: stats })
        );
    }

    private onHighLowStateEvent(event: HighLowStateEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const state = parser.state;
        const turn = parser.turn;
        const currentCard = parser.currentCard;
        const canAct = parser.canAct;
        const bet = parser.bet;
        const messages = parser.messages;

        console.log('[🎲] HighLow state received:', {
            state,
            turn,
            currentCard,
            canAct,
            bet,
            messages,
        });

        window.dispatchEvent(
            new CustomEvent('highlow_state', {
                detail: { state, turn, currentCard, canAct, bet, messages },
            })
        );
    }

    private onItemOfferPopupEvent(event: ItemOfferPopupEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const payload = {
            offerId: parser.offerId,
            senderId: parser.senderId,
            senderName: parser.senderName,
            itemName: parser.itemName,
            itemIcon: parser.itemIcon,
            itemCost: parser.itemCost,
        };

        window.dispatchEvent(
            new CustomEvent('item_offer_popup', { detail: payload })
        );

        //console.log('[📦] Received Item Offer Popup:', payload);
    }
    private onInspectUserStatsEvent(event: UserStatsEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const stats = parser.stats;
        window.dispatchEvent(
            new CustomEvent('user_inspect_stats', {
                detail: stats, // opponent stats
            })
        );
        //console.log('[📊] Inspect Stats Received User Stats:', stats);
    }

    private onUserLiveFeedEvent(event: UserLiveFeedEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const { username, figure, message, type, persistent } = parser.data;

        //console.log('[📡] Received LiveFeed packet:', parser.data);

        window.dispatchEvent(
            new CustomEvent('live_feed_event', {
                detail: { username, figure, message, type, persistent },
            })
        );
    }
    private onUpdateWorkStatusEvent(event: UpdateWorkStatusEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        // Dispatch to UI with working flag
        window.dispatchEvent(
            new CustomEvent('work_status_update', {
                detail: { isWorking: parser.isWorking },
            })
        );

        console.log(
            `🕒 Work status updated: ${
                parser.isWorking ? 'Working' : 'Not working'
            }`
        );
    }
    private onCancelGatherEvent(event: CancelGatherEvent): void {
        // Optional: only act if your packet sends a flag
        const parser = event.getParser();
        if (!parser) return;

        //console.log('⛔ Gathering was cancelled by server.');

        // Fire the cancellation event to your UI listener
        window.dispatchEvent(new CustomEvent('gather_cancel'));
    }
    private onOpenSellModuleEvent(event: OpenSellModuleEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        // Dispatch to client UI
        window.dispatchEvent(new CustomEvent('open_sell_module'));
        //console.log('🛒 Sell Module Triggered via OpenSellModuleEvent');
    }

    private onGangStatusEvent(event: GangStatusEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const data = parser.data;
        //console.log('[📣] GangStatusEvent -> inGang =', data.inGang);

        setTimeout(() => {
            // Open the correct UI
            window.dispatchEvent(
                new CustomEvent(
                    data.inGang ? 'open-gang-details' : 'open-create-gang',
                    { detail: data }
                )
            );

            // Dispatch gang_status_result with full data
            if (data.inGang) {
                window.dispatchEvent(
                    new CustomEvent('gang_status_result', { detail: data })
                );
            }
        }, 10);
    }

    private onMarketplaceModuleEvent(event: MarketplaceModuleEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const data = parser.data;
        console.log(
            '[🛒] MarketplaceModuleEvent -> listings =',
            data.listings?.length ?? 0
        );

        setTimeout(() => {
            // Always open the marketplace UI
            window.dispatchEvent(
                new CustomEvent('open-marketplace', { detail: data })
            );

            // Dispatch a more specific result event for anyone listening
            window.dispatchEvent(
                new CustomEvent('marketplace_data_result', { detail: data })
            );
        }, 10);
    }

    private onGangInviteEvent(event: GangInviteEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const inviterId: number = parser.inviterId;
        const inviterUsername: string = parser.inviterUsername;
        const gangName: string = parser.gangName;

        // Check user setting — default true if never set
        let invitesEnabled = true;
        try {
            const raw = localStorage.getItem('gangInvitesEnabled');
            if (raw !== null) invitesEnabled = JSON.parse(raw);
        } catch {}

        if (!invitesEnabled) {
           
            return; // don't show popup, don't send decline
        }

        // Fire popup if invites are enabled
        window.dispatchEvent(
            new CustomEvent('gang_invite_popup', {
                detail: {
                    inviter: inviterUsername,
                    inviterId: inviterId,
                    gangName: gangName,
                },
            })
        );
    }

    private onGangRanksEvent(event: GangRanksEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const ranks = parser.ranks.map((rank) => ({
            id: rank.id,
            name: rank.name,
            position: rank.position,
            canInvite: rank.canInvite,
            canKick: rank.canKick,
            canAlert: rank.canAlert,
            canEditRanks: rank.canEditRanks,
            canAccessBank: rank.canAccessBank,
            administrator: rank.administrator,
        }));
        //console.log('[onGangRanksEvent] recognized');
        // Dispatch to UI or state manager
        window.dispatchEvent(
            new CustomEvent('gang_ranks_received', {
                detail: {
                    ranks,
                },
            })
        );
    }
    private onGangMembersEvent(event: GangMembersEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const members = parser.members.map((m) => ({
            userId: m.userId,
            username: m.username,
            rankName: m.rankName,
            rankOrder: m.rankId, // use this for now
            figure: m.figure,
        }));

        //console.log('[📥] Gang members received:', members);

        window.dispatchEvent(
            new CustomEvent('gang_members_result', {
                detail: {
                    members,
                },
            })
        );
    }

    private onZeusStatsEvent(event: ZeusHealthEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const stats = parser.stats;

        //console.log('[⚡] Zeus stats received:', stats);

        window.dispatchEvent(
            new CustomEvent('zeus_stats_update', {
                detail: stats,
            })
        );
    }
    private onCorporationsListEvent(event: CorporationsListEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const corporations = parser.corporations;

        //console.log('[🏢] Corporations list received:', corporations);

        // 1) cache the latest payload so late subscribers can read it instantly
        (window as any).__olrpCache = {
            ...(window as any).__olrpCache,
            corporations,
            corporationsAt: Date.now(),
        };

        // 2) dispatch to any listeners that are already attached
        window.dispatchEvent(
            new CustomEvent('corporations_list_result', {
                detail: { corporations },
            })
        );
    }

    private onBlackjackStateEvent(event: BlackjackStateEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        // If your parser exposes .data (DTO), unpack it so the log + dispatch are explicit
        const {
            state,
            dealerReveal,
            dealerCards,
            dealerTotal,
            playerHand,
            playerTotal,
            playerTurn,
            canDouble,
            canSplit,
            insuranceOffered,
            bet,
            messages,
        } = parser.data ?? ({} as any);

        console.log('[🃏] Blackjack state received:', {
            state,
            dealerReveal,
            dealerCards,
            dealerTotal,
            playerHand,
            playerTotal,
            playerTurn,
            canDouble,
            canSplit,
            insuranceOffered,
            bet,
            messages,
        });

        window.dispatchEvent(
            new CustomEvent('blackjack_state', {
                detail: {
                    state,
                    dealerReveal,
                    dealerCards,
                    dealerTotal,
                    playerHand,
                    playerTotal,
                    playerTurn,
                    canDouble,
                    canSplit,
                    insuranceOffered,
                    bet,
                    messages,
                },
            })
        );
    }

    // --- Blackjack MODULE bridge ---
    private onBlackjackModuleEvent(event: BlackjackModuleEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const hasInfo = parser.hasInfo;
        const minBet = parser.minBet;
        const maxBet = parser.maxBet;

        //console.log('[🃏] Blackjack module info:', { hasInfo, minBet, maxBet });

        window.dispatchEvent(
            new CustomEvent('blackjack_module_result', {
                detail: { hasInfo, minBet, maxBet },
            })
        );
    }

    private onHighLowModuleEvent(event: HighLowModuleEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const hasInfo = parser.hasInfo;
        const minBet = parser.minBet;
        const maxBet = parser.maxBet;
        window.dispatchEvent(
            new CustomEvent('highlow_module_result', {
                detail: { hasInfo, minBet, maxBet },
            })
        );
    }

    private onSlotModuleEvent(event: SlotMachineModuleEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const hasInfo = parser.hasInfo;
        const minBet = parser.minBet; // expect 5
        const maxBet = parser.maxBet; // expect 25

        // Open your React UI (SlotView) the same way you do for High/Low:
        window.dispatchEvent(
            new CustomEvent('slotmachine_module_result', {
                detail: { hasInfo, minBet, maxBet },
            })
        );
    }
    private onWheelStateEvent = (event: WheelStateEvent): void => {
        const p = event.getParser();
        if (!p) return;

        window.dispatchEvent(
            new CustomEvent('bigwheel_state', {
                detail: {
                    state: p.state,
                    landedIndex: p.landedIndex,
                    landedFace: p.landedFace,
                    win: p.win,
                    betAmount: p.betAmount,
                    betFace: p.betFace,
                    faces: p.faces,
                },
            })
        );
    };

    private onSlotStateEvent(event: SlotMachineStateEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        // parser.state is already 'IDLE' | 'SPINNING' | 'ENDED'
        const stateStr = (parser.state || 'IDLE') as
            | 'IDLE'
            | 'SPINNING'
            | 'ENDED';

        const detail = {
            state: stateStr,
            reels: typeof parser.reels === 'number' ? parser.reels : 3,
            stops: Array.isArray(parser.stops) ? parser.stops : [],
            win: typeof parser.win === 'number' ? parser.win : 0,
            bet: typeof parser.bet === 'number' ? parser.bet : 0,
            messages: Array.isArray(parser.messages) ? parser.messages : [],
        };

        window.dispatchEvent(new CustomEvent('slotmachine_state', { detail }));
    }

    private onCorporationMembersEvent(event: CorporationMembersEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const members = parser.members;

        window.dispatchEvent(
            new CustomEvent('corporation_members_result', {
                detail: { members },
            })
        );
    }

    private onTimeOfDayEvent(event: TimeOfDayEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const data = parser.data; // { hhmm: "HH:mm", phase: "DAY" | "DUSK" | "NIGHT" | "DAWN" }

        window.dispatchEvent(
            new CustomEvent('time_of_day_update', {
                detail: data,
            })
        );
    }

    private onVirtualRoomInfoEvent(event: VirtualRoomInfoEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const data = parser.data; // { virtualRoomId, name, type, credits }

        window.dispatchEvent(
            new CustomEvent('virtual_room_info_update', {
                detail: data,
            })
        );
    }
    onCorporationRanksEvent = (event: CorporationRanksEvent): void => {
        const parser = event.getParser();
        if (!parser) return;

        const corporationId = parser.corporationId;
        const ranks = parser.ranks;

        // Optional: log
        // console.log('[🏷️] Corporation ranks received:', corporationId, ranks);

        // Dispatch to React via DOM bridge
        window.dispatchEvent(
            new CustomEvent('corporation_ranks_result', {
                detail: { corporationId, ranks },
            })
        );
    };
    private onPaletteHexesEvent(event: PaletteHexesEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const data = {
            part: parser.part, // e.g. "cc"
            paletteId: parser.paletteId, // e.g. 2
            hexes: parser.hexes, // string[] of "#RRGGBB"
        };

        window.dispatchEvent(
            new CustomEvent('palette_hexes_result', { detail: data })
        );
    }

    private onWheelModuleEvent(event: WheelModuleEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        window.dispatchEvent(
            new CustomEvent('bigwheel_module_result', {
                detail: {
                    hasInfo: parser.hasInfo,
                    faces: parser.faces, // string[] (order = clockwise)
                    payouts: parser.payouts, // { "1":1, "2":2, ... }
                    minBet: parser.minBet,
                    maxBet: parser.maxBet,
                    spinMs: parser.spinMs,
                },
            })
        );
    }

    private onWantedListEvent(event: WantedListEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const wantedUsers = parser.getWantedUsers();

        //console.log('[🚨] Wanted list received:', wantedUsers);

        window.dispatchEvent(
            new CustomEvent('wanted_list_update', {
                detail: wantedUsers,
            })
        );
    }

    private onHotelAlertEvent(event: HotelAlertEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const message = parser.message || '';

        //console.log('[🏨] Hotel alert received:', message);

        window.dispatchEvent(
            new CustomEvent('hotel_alert', {
                detail: { message },
            })
        );
    }

    private onInventoryItemsEvent(event: InventoryMessageEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const items: InventoryContext[] = parser.items.map((item: any) => ({
            id: item.id,
            item_id: item.item_id,
            user_id: item.user_id,
            quantity: item.quantity,
            durability: item.durability,
            slot: item.slot,
            last_used_at: item.last_used_at,
            name: item.name,
            description: item.description,
            item_type: item.item_type,
            cooldown_seconds: item.cooldown_seconds,
            damage_increase: item.damage_increase,
            health_increase: item.health_increase,
            durability_max: item.durability_max,
            rarity: item.rarity,
            icon_path: item.icon_path,
            max_stack: item.max_stack,
            created_at: item.created_at,
            stats: item.stats,
        }));

        console.log('Received Inventory:', items);

        InventoryStore.setItems(items); // ✅ this stores it globally

        // 🔥 Add this to update React component like SellModuleView
        window.dispatchEvent(new CustomEvent('inventory_update'));
    }

    private onConnectionOpenedEvent(event: Event): void {
        this.logger.log('Connection Initialized');
    }

    private onConnectionClosedEvent(event: CloseEvent): void {
        this.logger.log('Connection Closed');
    }

    private onConnectionErrorEvent(event: Event): void {
        this.logger.log('Connection Error');
    }

    private onConnectionAuthenticatedEvent(event: NitroEvent): void {
        this.logger.log('Connection Authenticated');
        console.log('onConnectAuthenticatedEvent has been registered');

        if (this._connection) {
            this._connection.authenticated();

            this.registerMessageEvent(
                new TimeOfDayEvent(this.onTimeOfDayEvent.bind(this))
            );
            this.registerMessageEvent(
                new WheelStateEvent(this.onWheelStateEvent.bind(this))
            );

            this.registerMessageEvent(
                new HighLowModuleEvent(this.onHighLowModuleEvent.bind(this))
            );

            this.registerMessageEvent(
                new HighLowStateEvent(this.onHighLowStateEvent.bind(this))
            );

            this.registerMessageEvent(
                new GangRanksEvent(this.onGangRanksEvent.bind(this))
            );
            // Now it's safe to register message event
            this.registerMessageEvent(
                new GangStatusEvent(this.onGangStatusEvent.bind(this))
            );
            // console.log("Registered gangstatusevent");
            this.registerMessageEvent(
                new InspectUserStatsEvent(
                    this.onInspectUserStatsEvent.bind(this)
                )
            );
            //console.log('Registered InspectUserStatsEvent');
            this.registerMessageEvent(
                new CorporationsListEvent(
                    this.onCorporationsListEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new MarketplaceModuleEvent(
                    this.onMarketplaceModuleEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new CorporationRanksEvent(
                    this.onCorporationRanksEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new WheelModuleEvent(this.onWheelModuleEvent.bind(this))
            );
            this.registerMessageEvent(
                new ItemOfferPopupEvent(this.onItemOfferPopupEvent.bind(this))
            );
            this.registerMessageEvent(
                new WantedListEvent(this.onWantedListEvent.bind(this))
            );

            this.registerMessageEvent(
                new UpdateWorkStatusEvent(
                    this.onUpdateWorkStatusEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new CorporationMembersEvent(
                    this.onCorporationMembersEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new GangMembersEvent(this.onGangMembersEvent.bind(this))
            );
            this.registerMessageEvent(
                new HotelAlertEvent(this.onHotelAlertEvent.bind(this))
            );
            this.registerMessageEvent(
                new SlotMachineStateEvent(this.onSlotStateEvent.bind(this))
            );
            this.registerMessageEvent(
                new SlotMachineModuleEvent(this.onSlotModuleEvent.bind(this))
            );
            this.registerMessageEvent(
                new PaletteHexesEvent(this.onPaletteHexesEvent.bind(this))
            );
            this.registerMessageEvent(
                new EquippedLoadoutEvent(this.onEquippedLoadoutEvent.bind(this))
            );
            this.registerMessageEvent(
                new VirtualRoomInfoEvent(this.onVirtualRoomInfoEvent.bind(this))
            );
            this.registerMessageEvent(
                new OpenSellModuleEvent(this.onOpenSellModuleEvent.bind(this))
            );
            this.registerMessageEvent(
                new CancelGatherEvent(this.onCancelGatherEvent.bind(this))
            );
            this.registerMessageEvent(
                new GatherProgressEvent(this.onGatherProgressEvent.bind(this))
            );
            this.registerMessageEvent(
                new InventoryMessageEvent(this.onInventoryItemsEvent.bind(this))
            );
            this.registerMessageEvent(
                new GangInviteEvent(this.onGangInviteEvent.bind(this))
            );
            //console.log('Registered InventoryMessageEvent');

            this.registerMessageEvent(
                new UserStatsEvent(this.onUserStatsEvent.bind(this))
            );
            //console.log('Registered UserStatsEvent');
            this.registerMessageEvent(
                new ZeusHealthEvent(this.onZeusStatsEvent.bind(this))
            );
            //console.log('Registered ZeusStatsEvent');
            this.registerMessageEvent(
                new UserLiveFeedEvent(this.onUserLiveFeedEvent.bind(this))
            );

            this.registerMessageEvent(
                new BlackjackModuleEvent(this.onBlackjackModuleEvent.bind(this))
            );

            this.registerMessageEvent(
                new BlackjackStateEvent(this.onBlackjackStateEvent.bind(this))
            );
        }
    }

    public connectionInit(socketUrl: string): void {
        this.logger.log(`Initializing Connection: ${socketUrl}`);
    }

    public registerMessageEvent(event: IMessageEvent): IMessageEvent {
        if (this._connection) this._connection.addMessageEvent(event);
        //console.log("Recieved header:", event)

        return event;
    }

    public removeMessageEvent(event: IMessageEvent): void {
        if (!this._connection) return;

        this._connection.removeMessageEvent(event);
    }

    public get demo(): NitroCommunicationDemo {
        return this._demo;
    }

    public get connection(): IConnection {
        return this._connection;
    }
}
