// @nitrots/nitro-renderer/src/nitro/communication/NitroCommunicationManager.ts
import { InventoryMessageEvent } from '@nitrots/nitro-renderer';
import { CancelGatherEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/CancelGatherEvent';
import { CasinoJackpotUpdateEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/CasinoJackpotUpdateEvent';
import { CorporationMembersEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/CorporationMembersEvent';
import { CorporationRanksEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/CorporationRanksEvent';
import { EquippedLoadoutEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/EquippedLoadoutEvent';
import { GangInviteEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/GangInviteEvent';
import { GangMembersEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/GangMembersEvent';
import { GangStatusEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/GangStatusEvent';
import { HeistTimerEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/HeistTimerEvent';
import { HeistTimerStopEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/HeistTimerStopEvent';
import { ItemOfferPopupEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/ItemOfferPopupEvent';
import { MarketplaceModuleEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/MarketplaceModuleEvent';
import { MythicalBlueprintOpenEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/MythicalBlueprintOpenEvent';
import { MythicalBlueprintResultEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/MythicalBlueprintResultEvent';
import { MythicalBlueprintUpdateEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/MythicalBlueprintUpdateEvent';
import { PaletteHexesEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/PaletteHexesEvent';
import { ArenaLeaderboardEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/ArenaLeaderboardEvent';
import { ArenaStatusMessageEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/ArenaStatusMessageEvent';
import { ATMOpenMessageEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/ATMOpenMessageEvent';
import { BlackjackModuleEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/BlackjackModuleEvent';
import { BlackjackStateEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/BlackjackStateEvent';
import { ChatBubblePurchaseOkEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/ChatBubblePurchaseOkEvent';
import { ChatBubbleShopEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/ChatBubbleShopEvent';
import { ClickThroughStateEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/ClickThroughStateEvent';
import { CommandsListMessageEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/CommandsListMessageEvent';
import { CorporationsListEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/CorporationsListEvent';
import { DiscordVerificationRequiredEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/DiscordVerificationRequiredEvent';
import { DoorDashOrdersEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/DoorDashOrdersEvent';
import { GangRanksEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/GangRanksEvent';
import { GatherProgressEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/GatherProgressEvent';
import { HaircutStylesOfferEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/HaircutStylesOfferEvent';
import { HighLowModuleEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/HighLowModuleEvent';
import { HighLowStateEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/HighLowStateEvent';
import { HotelAlertEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/HotelAlertEvent';
import { HotRoomsListMessageEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/HotRoomsListMessageEvent';
import { InspectUserStatsEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/InspectUserStatsEvent';
import { JukeboxOpenEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/JukeboxOpenEvent';
import { JukeboxStateEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/JukeboxStateEvent';
import { JukeboxToastEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/JukeboxToastEvent';
import { NameIconPurchaseOkEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/NameIconPurchaseOkEvent';
import NameIconShopEvent from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/NameIconShopEvent';
import { OpenAvatarEditorEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/OpenAvatarEditorEvent';
import { OpenCorpJobBoardEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/OpenCorpJobBoardEvent';
import { OpenFishingModuleMessageEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/OpenFishingModuleEvent';
import { OpenSellModuleEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/OpenSellModuleEvent';
import { OpponentStatsEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/OpponentStatsEvent';
import { PoliceCallEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/PoliceCallEvent';
import { ProfileAchievementsEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/ProfileAchievementsEvent';
import { QuickGuideEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/QuickGuideEvent';
import { SlotMachineModuleEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/SlotMachineModuleEvent';
import { SlotMachineStateEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/SlotMachineStateEvent';
import { TaxiOpenListEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/TaxiOpenListEvent';
import { TimeOfDayEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/TimeOfDayEvent';
import { TokenPurchaseOkEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/TokenPurchaseOkEvent';
import { TokensStoreEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/TokensStoreEvent';
import { UpdateWorkStatusEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/UpdateWorkStatusEvent';
import { UserStatsEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/UserStatsEvent';
import { VirtualRoomInfoEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/VirtualRoomInfoEvent';
import { WantedListEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/WantedListEvent';
import { WheelModuleEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/WheelModuleEvent';
import { WheelStateEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/WheelStateEvent';
import { ZeusHealthEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/ZeusHealthEvent';
import { RoleplayCombatCooldownEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/RoleplayCombatCooldownEvent';
import { TradeCloseEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/TradeCloseEvent';
import { TradeOpenEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/TradeOpenEvent';
import { TradeResultEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/TradeResultEvent';
import { TradeUpdateEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/TradeUpdateEvent';
import { TurfCaptureProgressEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/TurfCaptureProgressEvent';
import { UserLiveFeedEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/UserLiveFeedEvent';
import { InventoryStore } from '@nitrots/nitro-renderer/src/nitro/communication/messages/parser/roleplay/InventoryStore';
import { InventoryContext } from '../../../../../../src/api/contexts/inventory/InventoryContext';
import { IMessageEvent } from '../../core';
import { NitroManager } from '../../core/common/NitroManager';
import { IConnection } from '../../core/communication/connections/IConnection';
import { IConnectionStateListener } from '../../core/communication/connections/IConnectionStateListener';
import { SocketConnectionEvent } from '../../core/communication/events/SocketConnectionEvent';
import { ICommunicationManager } from '../../core/communication/ICommunicationManager';
import { IMessageConfiguration } from '../../core/communication/messages/IMessageConfiguration';
import { NitroEvent } from '../../core/events/NitroEvent';
import { Nitro } from '../Nitro';
import { NitroCommunicationDemo } from './demo/NitroCommunicationDemo';
import { NitroCommunicationDemoEvent } from './demo/NitroCommunicationDemoEvent';
import { INitroCommunicationManager } from './INitroCommunicationManager';
import { NameIconEquippedEvent } from '@nitrots/nitro-renderer/src/nitro/communication/messages/incoming/roleplay/NameIconEquippedEvent';
import { NitroMessages } from './NitroMessages';
const TURF_TO_VROOM: Record<string, number> = {
    NORTH: 18,
    SOUTH: 12,
    EAST: 3,
    WEST: 4,
};

export class NitroCommunicationManager
    extends NitroManager
    implements INitroCommunicationManager, IConnectionStateListener
{
    private _communication: ICommunicationManager;
    private _connection: IConnection;
    private _messages: IMessageConfiguration;
private _lastOpponentHealthByUserId: Map<number, number> = new Map();
    private _demo: NitroCommunicationDemo;

    constructor(communication: ICommunicationManager) {
        super();

        this._communication = communication;
        this._connection = null;
        this._messages = new NitroMessages();

        this._demo = new NitroCommunicationDemo(this);

        this.onConnectionOpenedEvent = this.onConnectionOpenedEvent.bind(this);
        this.onConnectionClosedEvent = this.onConnectionClosedEvent.bind(this);
        this.onConnectionErrorEvent = this.onConnectionErrorEvent.bind(this);
        this.onConnectionAuthenticatedEvent =
            this.onConnectionAuthenticatedEvent.bind(this);
    }

    protected onInit(): void {
        if (this._connection) return;
        console.log('OnIt from NitroCommunicationManager initialized!!');

        Nitro.instance.events.addEventListener(
            NitroCommunicationDemoEvent.CONNECTION_AUTHENTICATED,
            this.onConnectionAuthenticatedEvent
        );

        this._connection = this._communication.createConnection(this);

        this._connection.registerMessages(this._messages);

        this._connection.addEventListener(
            SocketConnectionEvent.CONNECTION_OPENED,
            this.onConnectionOpenedEvent
        );
        this._connection.addEventListener(
            SocketConnectionEvent.CONNECTION_CLOSED,
            this.onConnectionClosedEvent
        );
        this._connection.addEventListener(
            SocketConnectionEvent.CONNECTION_ERROR,
            this.onConnectionErrorEvent
        );

        if (this._demo) this._demo.init();

        this._connection.init(
            Nitro.instance.getConfiguration<string>('socket.url')
        );
    }

    protected onDispose(): void {
        if (this._demo) this._demo.dispose();

        if (this._connection) {
            this._connection.removeEventListener(
                SocketConnectionEvent.CONNECTION_OPENED,
                this.onConnectionOpenedEvent
            );
            this._connection.removeEventListener(
                SocketConnectionEvent.CONNECTION_CLOSED,
                this.onConnectionClosedEvent
            );
            this._connection.removeEventListener(
                SocketConnectionEvent.CONNECTION_ERROR,
                this.onConnectionErrorEvent
            );
        }

        Nitro.instance.events.removeEventListener(
            NitroCommunicationDemoEvent.CONNECTION_AUTHENTICATED,
            this.onConnectionAuthenticatedEvent
        );

        super.onDispose();
    }

    private onGatherProgressEvent(event: GatherProgressEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const duration = parser.duration;
        const name = parser.itemName || 'Gathering...';
        const icon = parser.icon || '';

        console.log(
            `[📦] GatherProgressEvent → duration=${duration} name=${name} icon=${icon}`
        );

        // Dispatch to show or update the progress bar
        window.dispatchEvent(
            new CustomEvent('gather_progress', {
                detail: { duration, name, icon },
            })
        );
    }

    private onCasinoJackpotUpdateEvent(event: CasinoJackpotUpdateEvent): void {
const parser = event.getParser();
if (!parser) return;

const amount = parser.amount || 0;

console.log(`[🎰] CasinoJackpotUpdateEvent → amount=${amount}`);

window.dispatchEvent(
new CustomEvent('casino_jackpot_update', {
detail: { amount }
})
);
}

private onRoleplayCombatCooldownEvent(event: RoleplayCombatCooldownEvent): void
{
const parser = event.getParser();
if (!parser) return;

const weaponType = parser.weaponType ?? 4;
const totalMs = parser.totalMs ?? 0;
const remainingMs = parser.remainingMs ?? 0;

console.log(`[⚔️] RoleplayCombatCooldownEvent → weaponType=${weaponType} totalMs=${totalMs} remainingMs=${remainingMs}`);

window.dispatchEvent(
new CustomEvent('combat_cooldown', {
detail: { weaponType, totalMs, remainingMs }
})
);
}

private onNameIconEquippedEvent(event: NameIconEquippedEvent): void
{
    const parser = event.getParser();

    if(!parser) return;

    const userId = parser.userId ?? 0;
    const iconId = parser.iconId ?? 0;
    const iconKey = parser.iconKey ?? '';

    console.log(`[🏷️] NameIconEquippedEvent → userId=${ userId } iconId=${ iconId } iconKey=${ iconKey }`);

    window.dispatchEvent(
        new CustomEvent('name_icon_equipped_for_user', {
            detail: { userId, iconId, iconKey }
        })
    );
}


    private onEquippedLoadoutEvent(event: EquippedLoadoutEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const payload = {
            weapon: parser.weapon, // or null
            shield: parser.shield, // or null
            potion: parser.potion, // or null
        };

        // Dispatch to client UI
        window.dispatchEvent(
            new CustomEvent('equipped_loadout_update', { detail: payload })
        );
        console.log('🧰 EquippedLoadoutEvent ->', payload);
    }
    private onUserStatsEvent(event: UserStatsEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const s = parser.stats;

        // 1) Single canonical broadcast for the whole app
        window.dispatchEvent(new CustomEvent('user_stats', { detail: s }));

        // 2) Legacy/feature-specific broadcast (if other UI listens to this)
        window.dispatchEvent(
            new CustomEvent('user_stats_update', { detail: s })
        );

        // 3) 💰 Keep ATM in sync immediately on login/respawn/etc.
        window.dispatchEvent(
            new CustomEvent('atm_update', {
                detail: {
                    username: s.username ?? '',
                    bank: Number.isFinite(s.bank) ? s.bank : 0,
                    cash: Number.isFinite(s.cash) ? s.cash : 0,
                },
            })
        );
    }

    private onHighLowStateEvent(event: HighLowStateEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const state = parser.state;
        const turn = parser.turn;
        const currentCard = parser.currentCard;
        const canAct = parser.canAct;
        const bet = parser.bet;
        const messages = parser.messages;

        console.log('[🎲] HighLow state received:', {
            state,
            turn,
            currentCard,
            canAct,
            bet,
            messages,
        });

        window.dispatchEvent(
            new CustomEvent('highlow_state', {
                detail: { state, turn, currentCard, canAct, bet, messages },
            })
        );
    }

    private onItemOfferPopupEvent(event: ItemOfferPopupEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const payload = {
            offerId: parser.offerId,
            senderId: parser.senderId,
            senderName: parser.senderName,
            itemName: parser.itemName,
            itemIcon: parser.itemIcon,
            itemCost: parser.itemCost,
        };

        window.dispatchEvent(
            new CustomEvent('item_offer_popup', { detail: payload })
        );

        //console.log('[📦] Received Item Offer Popup:', payload);
    }

    private onArenaStatusEvent(event: ArenaStatusMessageEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const hasSelf = parser.selfId > 0 && !!parser.selfUsername;
        const hasOpponent = parser.oppId > 0 && !!parser.oppUsername;

        const payload = {
            inQueue: parser.inQueue,
            isMatched: parser.isMatched,

            self: hasSelf
                ? {
                      userId: parser.selfId,
                      username: parser.selfUsername,
                      figure: parser.selfFigure,
                      level: parser.selfLevel,
                      wins: parser.selfWins,
                      losses: parser.selfLosses,
                  }
                : null,

            // opponent is present whenever oppId > 0, match state is still
            // decided by inQueue + isMatched on the React side
            opponent: hasOpponent
                ? {
                      userId: parser.oppId,
                      username: parser.oppUsername,
                      figure: parser.oppFigure,
                      level: parser.oppLevel,
                      wins: parser.oppWins,
                      losses: parser.oppLosses,
                  }
                : null,
        };

        window.dispatchEvent(
            new CustomEvent('arena_status', { detail: payload })
        );

        // console.log('[⚔️] Arena status update:', payload);
    }
    private onDoorDashOrdersEvent(event: DoorDashOrdersEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        console.log(
            '[Bridge] DoorDashOrdersEvent fired, orders =',
            parser.orders
        );

        window.dispatchEvent(
            new CustomEvent('doordash_orders', {
                detail: parser.orders,
            })
        );
    }

    private onATMOpenEvent(event: ATMOpenMessageEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const payload = {
            username: parser.username, // 🧍 new field
            cash: parser.cash, // 💵 on-hand credits
            bank: parser.bank, // 🏦 bank credits
        };

        window.dispatchEvent(
            new CustomEvent('atm_open', {
                detail: payload, // full ATM payload
            })
        );

        // console.log('[🏧] ATM Open Received:', payload);
    }
    private onArenaLeaderboardEvent(event: ArenaLeaderboardEvent): void
{
const parser = event.getParser();
if (!parser) return;

const entries = parser.entries;

window.dispatchEvent(new CustomEvent('arena_leaderboard', {
detail: {
entries
}
}));

// console.log('[🏆] Arena Leaderboard Received:', entries);
}
    private onInspectUserStatsEvent(event: UserStatsEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const stats = parser.stats;
        window.dispatchEvent(
            new CustomEvent('user_inspect_stats', {
                detail: stats, // opponent stats
            })
        );
        //console.log('[📊] Inspect Stats Received User Stats:', stats);
    }

    private onTokensStoreEvent(event: TokensStoreEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        window.dispatchEvent(
            new CustomEvent('store_tokens_shop', {
                detail: {
                    diamonds: parser.diamonds,
                    items: parser.items,
                },
            })
        );
        // console.log('[🛍️] Tokens store payload', parser.items, '💎', parser.diamonds);
    }

    private onTokenPurchaseOkEvent(event: TokenPurchaseOkEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        window.dispatchEvent(
            new CustomEvent('store_tokens_purchase_ok_raw', {
                detail: {
                    tokenId: parser.tokenId,
                    diamonds: parser.diamonds,
                    activeUntilTs: parser.activeUntilTs,
                },
            })
        );
        // console.log('[✅] Token purchase OK', parser.tokenId, '💎', parser.diamonds, 'until', parser.activeUntilTs);
    }

    private onUserLiveFeedEvent(event: UserLiveFeedEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const { username, figure, message, type, persistent } = parser.data;

        //console.log('[📡] Received LiveFeed packet:', parser.data);

        window.dispatchEvent(
            new CustomEvent('live_feed_event', {
                detail: { username, figure, message, type, persistent },
            })
        );
    }

   private _policeCallDedup: Map<string, number> = new Map();

private onPoliceCallEvent(event: PoliceCallEvent): void {
const parser = event.getParser();
if (!parser) return;

const { username, figure, message, virtualRoomId, virtualRoomName } = parser.data;

// --- dedupe (in case packet fires twice) ---
const now = Date.now();
const key = `${username}|${virtualRoomId}|${virtualRoomName}|${message}`;

const last = this._policeCallDedup.get(key) || 0;
if (now - last < 1500) return;

this._policeCallDedup.set(key, now);

// cleanup old keys
if (this._policeCallDedup.size > 200) {
for (const [k, t] of this._policeCallDedup.entries()) {
if (now - t > 60_000) this._policeCallDedup.delete(k);
}
}

// ✅ ONLY this event — used by your PoliceCallView and/or bridge logic
window.dispatchEvent(
new CustomEvent("police_call_received", {
detail: {
username,
figure,
message: String(message ?? ""),
virtualRoomId: Number(virtualRoomId ?? 0),
virtualRoomName: String(virtualRoomName ?? ""),
},
})
);
}


    private onUpdateWorkStatusEvent(event: UpdateWorkStatusEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        // Dispatch to UI with working flag
        window.dispatchEvent(
            new CustomEvent('work_status_update', {
                detail: { isWorking: parser.isWorking },
            })
        );

        console.log(
            `🕒 Work status updated: ${
                parser.isWorking ? 'Working' : 'Not working'
            }`
        );
    }
    private onCancelGatherEvent(event: CancelGatherEvent): void {
        // Optional: only act if your packet sends a flag
        const parser = event.getParser();
        if (!parser) return;

        //console.log('⛔ Gathering was cancelled by server.');

        // Fire the cancellation event to your UI listener
        window.dispatchEvent(new CustomEvent('gather_cancel'));
    }
    private onOpenSellModuleEvent(event: OpenSellModuleEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        // Dispatch to client UI
        window.dispatchEvent(new CustomEvent('open_sell_module'));
        //console.log('🛒 Sell Module Triggered via OpenSellModuleEvent');
    }

    private onGangStatusEvent(event: GangStatusEvent): void {
        try {
            const parser = event.getParser?.();
            if (!parser) return;

            const data = parser.data;

            // Persist for instant UI hydration on next open
            (window as any).__lastGangStatus = data;

            // Always tell the UI the latest status (even if not in a gang) so it can clear state
            window.dispatchEvent(
                new CustomEvent('gang_status_result', { detail: data })
            );
            window.dispatchEvent(
                new CustomEvent('gang_status_received', { detail: data })
            );

            // Small microtask delay lets listeners attach first
            setTimeout(() => {
                // Close both views first (listeners should noop if already closed)
                window.dispatchEvent(new CustomEvent('close-gang-details'));
                window.dispatchEvent(new CustomEvent('close-create-gang'));

                // Open the appropriate view
                window.dispatchEvent(
                    new CustomEvent(
                        data.inGang ? 'open-gang-details' : 'open-create-gang',
                        { detail: data }
                    )
                );

                // Optional: if in gang, proactively refresh members/ranks
                if (data.inGang) {
                    window.dispatchEvent(
                        new CustomEvent('request_gang_members')
                    );
                    window.dispatchEvent(new CustomEvent('request_gang_ranks'));
                }
            }, 10);
        } catch (err) {
            console.error('[GangBridge] onGangStatusEvent error:', err);
        }
    }

    private onMarketplaceModuleEvent(event: MarketplaceModuleEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const data = parser.data;
        console.log(
            '[🛒] MarketplaceModuleEvent -> listings =',
            data.listings?.length ?? 0
        );

        setTimeout(() => {
            // Always open the marketplace UI
            window.dispatchEvent(
                new CustomEvent('open-marketplace', { detail: data })
            );

            // Dispatch a more specific result event for anyone listening
            window.dispatchEvent(
                new CustomEvent('marketplace_data_result', { detail: data })
            );
        }, 10);
    }

    private onGangInviteEvent(event: GangInviteEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const inviterId: number = parser.inviterId;
        const inviterUsername: string = parser.inviterUsername;
        const gangName: string = parser.gangName;

        // Check user setting — default true if never set
        let invitesEnabled = true;
        try {
            const raw = localStorage.getItem('gangInvitesEnabled');
            if (raw !== null) invitesEnabled = JSON.parse(raw);
        } catch {}

        if (!invitesEnabled) {
            return; // don't show popup, don't send decline
        }

        // Fire popup if invites are enabled
        window.dispatchEvent(
            new CustomEvent('gang_invite_popup', {
                detail: {
                    inviter: inviterUsername,
                    inviterId: inviterId,
                    gangName: gangName,
                },
            })
        );
    }

  private onProfileAchievementsEvent(event: ProfileAchievementsEvent): void
{
const parser = event.getParser();
if (!parser) return;

const achievements = parser.achievements;
const userId = parser.userId; // ✅ IMPORTANT

// Bridge to React/UI
window.dispatchEvent(
new CustomEvent('profile_achievements_update', {
detail: { userId, achievements },
})
);

console.log('[Profile] Achievements update received for user:', userId);
}
    

    private onChatBubbleShopEvent(event: ChatBubbleShopEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        // Convert to the object shape SettingsView expects
        window.dispatchEvent(
            new CustomEvent('chat_bubble_shop', {
                detail: {
                    items: parser.items.map((i) => ({
                        bubbleId: i.id,
                        price: i.price,
                        owned: i.owned,
                        equipped: i.equipped,
                    })),
                    diamonds: parser.diamonds, // optional (can be undefined)
                    equippedId: parser.equippedId, // helps select active bubble
                },
            })
        );

        console.log(
            `[🗯] Dispatched chat_bubble_shop with ${parser.items.length} items (equippedId=${parser.equippedId})`
        );
    }

    
   
    

    private onOpenAvatarEditorEvent(event: OpenAvatarEditorEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        // mode = 0 or 1 depending on server. Safe fallback: 0
        const mode = parser.mode ?? 0;

        console.log('[Bridge] OpenAvatarEditorEvent received. Mode:', mode);

        // 🔥 Trigger DOM event that AvatarEditorView.tsx listens for
        window.dispatchEvent(
            new CustomEvent('open_avatar_editor', {
                detail: { mode },
            })
        );
    }

    private onTradeOpenEvent(event: TradeOpenEvent): void
{
const parser = event.getParser();
if(!parser) return;

console.log('[Bridge] TradeOpenEvent received.', {
sessionId: parser.sessionId,
userAId: parser.userAId,
userBId: parser.userBId,
usernameA: parser.usernameA,
usernameB: parser.usernameB,

});

window.dispatchEvent(new CustomEvent('trade_open', {
detail: {
sessionId: parser.sessionId,
userAId: parser.userAId,
userBId: parser.userBId,
usernameA: parser.usernameA,
usernameB: parser.usernameB,

}
}));
}

private onMythicalBlueprintOpenEvent(event: MythicalBlueprintOpenEvent): void
{
const parser = event.getParser();
if(!parser) return;

console.log('[Bridge] MythicalBlueprintOpenEvent received.', {
sessionId: parser.sessionId,
blueprintItemId: parser.blueprintItemId,
slotCount: parser.slotCount,
weaponInvItemId: parser.weaponInvItemId,
augmentInvItemId: parser.augmentInvItemId,
optionalInvItemId: parser.optionalInvItemId,
predictedUpgradeId: parser.predictedUpgradeId,
craftable: parser.craftable
});

window.dispatchEvent(new CustomEvent('mythical_blueprint_open', {
detail: {
sessionId: parser.sessionId,
blueprintItemId: parser.blueprintItemId,
slotCount: parser.slotCount,
weaponInvItemId: parser.weaponInvItemId,
augmentInvItemId: parser.augmentInvItemId,
optionalInvItemId: parser.optionalInvItemId,
predictedUpgradeId: parser.predictedUpgradeId,
craftable: parser.craftable
}
}));
}


private onMythicalBlueprintUpdateEvent(event: MythicalBlueprintUpdateEvent): void
{
const parser = event.getParser();
if(!parser) return;

console.log('[Bridge] MythicalBlueprintUpdateEvent received.', {
sessionId: parser.sessionId,
blueprintItemId: parser.blueprintItemId,
slotCount: parser.slotCount,
weaponInvItemId: parser.weaponInvItemId,
augmentInvItemId: parser.augmentInvItemId,
optionalInvItemId: parser.optionalInvItemId,
predictedUpgradeId: parser.predictedUpgradeId,
craftable: parser.craftable
});

window.dispatchEvent(new CustomEvent('mythical_blueprint_update', {
detail: {
sessionId: parser.sessionId,
blueprintItemId: parser.blueprintItemId,
slotCount: parser.slotCount,
weaponInvItemId: parser.weaponInvItemId,
augmentInvItemId: parser.augmentInvItemId,
optionalInvItemId: parser.optionalInvItemId,
predictedUpgradeId: parser.predictedUpgradeId,
craftable: parser.craftable
}
}));
}
 

private onMythicalBlueprintResultEvent(event: MythicalBlueprintResultEvent): void
{
const parser = event.getParser();
if(!parser) return;

console.log('[Bridge] MythicalBlueprintResultEvent received.', {
success: parser.success,
message: parser.message,
appliedUpgradeId: parser.appliedUpgradeId
});

window.dispatchEvent(new CustomEvent('mythical_blueprint_result', {
detail: {
success: parser.success,
message: parser.message,
appliedUpgradeId: parser.appliedUpgradeId
}
}));
}




private onTradeUpdateEvent(event: TradeUpdateEvent): void
{
const parser = event.getParser();
if(!parser) return;

console.log('[Bridge] TradeUpdateEvent received.', {
sessionId: parser.sessionId,
aConfirmed: parser.aConfirmed,
bConfirmed: parser.bConfirmed,
aSlots: parser.aSlots?.length,
bSlots: parser.bSlots?.length
});

window.dispatchEvent(new CustomEvent('trade_update', {
detail: {
sessionId: parser.sessionId,
aConfirmed: parser.aConfirmed,
bConfirmed: parser.bConfirmed,
aSlots: parser.aSlots,
bSlots: parser.bSlots
}
}));
}



private onTradeResultEvent(event: TradeResultEvent): void
{
const parser = event.getParser();
if(!parser) return;

console.log('[Bridge] TradeResultEvent received.', {
success: parser.success,
message: parser.message
});

window.dispatchEvent(new CustomEvent('trade_result', {
detail: {
success: parser.success,
message: parser.message
}
}));
}


private onTradeCloseEvent(event: TradeCloseEvent): void
{
const parser = event.getParser();
if(!parser) return;

console.log('[Bridge] TradeCloseEvent received.', {
sessionId: parser.sessionId
});

window.dispatchEvent(new CustomEvent('trade_close', {
detail: {
sessionId: parser.sessionId
}
}));
}






 


    private onChatBubblePurchaseOkEvent(
        event: ChatBubblePurchaseOkEvent
    ): void {
        const parser = event.getParser();
        if (!parser) return;

        // Shape DiamondsStoreView expects
        window.dispatchEvent(
            new CustomEvent('chat_bubble_purchase_ok_raw', {
                detail: {
                    bubbleId: parser.bubbleId,
                    diamonds: parser.diamonds,
                },
            })
        );

        console.log(
            `[✅] Dispatched chat_bubble_purchase_ok_raw (bubbleId=${parser.bubbleId}, diamonds=${parser.diamonds})`
        );
    }

    private onNameIconShopEvent(event: NameIconShopEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        window.dispatchEvent(
            new CustomEvent('name_icon_shop', {
                detail: {
                    items: parser.items.map((i) => ({
                        iconId: i.id,
                        key: i.key,
                        price: i.price,
                        owned: i.owned,
                        equipped: i.equipped,
                    })),
                    equippedId: parser.equippedId,
                },
            })
        );

        console.log(
            `[🔷] Dispatched name_icon_shop with ${parser.items.length} items (equippedId=${parser.equippedId})`
        );
    }

    private onTaxiOpenListEvent(event: TaxiOpenListEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        // TaxiView (new format) expects: { currentRoomId, currentVrid, destinations: [{ roomId, virtualRoomId, virtualName, userCount?, thumbnailUrl?, roomName? }] }
        const destinations = parser.destinations.map((d: any) => ({
            roomId: Number(d.roomId),
            virtualRoomId: Number(d.virtualRoomId),
            virtualName: String(d.virtualName || ''), // <- key name must be virtualName
            userCount:
                typeof d.userCount === 'number' ? d.userCount : undefined,
            thumbnailUrl: d.thumbnailUrl ?? undefined, // optional
            roomName: d.roomName ?? undefined, // optional
        }));

        window.dispatchEvent(
            new CustomEvent('taxi_open_list', {
                detail: {
                    currentRoomId: Number(parser.currentRoomId || 0),
                    currentVrid: Number(parser.currentVrid || 0),
                    destinations,
                },
            })
        );
    }
    private onGangRanksEvent(event: GangRanksEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const ranks = parser.ranks.map((rank) => ({
            id: rank.id,
            name: rank.name,
            position: rank.position,
            canInvite: rank.canInvite,
            canKick: rank.canKick,
            canAlert: rank.canAlert,
            canEditRanks: rank.canEditRanks,
            canAccessBank: rank.canAccessBank,
            administrator: rank.administrator,
        }));
        //console.log('[onGangRanksEvent] recognized');
        // Dispatch to UI or state manager
        window.dispatchEvent(
            new CustomEvent('gang_ranks_received', {
                detail: {
                    ranks,
                },
            })
        );
    }
private onGangMembersEvent(event: GangMembersEvent): void {
const parser = event.getParser();
if (!parser) return;

const members = parser.members.map((m) => ({
userId: m.userId,
username: m.username,
rankId: m.rankId,
rankName: m.rankName,
figure: m.figure,
}));

window.dispatchEvent(
new CustomEvent('gang_members_result', {
detail: { members }
})
);

window.dispatchEvent(
new CustomEvent('gang_members_received', {
detail: { members }
})
);
}

    private onZeusStatsEvent(event: ZeusHealthEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const stats = parser.stats;

        //console.log('[⚡] Zeus stats received:', stats);

        window.dispatchEvent(
            new CustomEvent('zeus_stats_update', {
                detail: stats,
            })
        );
    }

    private onHeistTimerEvent(event: HeistTimerEvent): void {
const parser = event.getParser();
if (!parser) return;

const payload = {
roomId: parser.roomId,
gangId: parser.gangId,
remainingSeconds: parser.remainingSeconds,
totalSeconds: parser.totalSeconds
};

window.dispatchEvent(
new CustomEvent('heist_timer_start', {
detail: payload
})
);

// console.log('[HeistTimer] start/update received:', payload);
}

private onHeistTimerStopEvent(event: HeistTimerStopEvent): void {
const parser = event.getParser();
if (!parser) return;

window.dispatchEvent(
new CustomEvent('heist_timer_stop', {
detail: { roomId: parser.roomId }
})
);

// console.log('[HeistTimer] stop received:', parser.roomId);
}

    private onCorporationsListEvent(event: CorporationsListEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const corporations = parser.corporations;

        //console.log('[🏢] Corporations list received:', corporations);

        // 1) cache the latest payload so late subscribers can read it instantly
        (window as any).__olrpCache = {
            ...(window as any).__olrpCache,
            corporations,
            corporationsAt: Date.now(),
        };

        // 2) dispatch to any listeners that are already attached
        window.dispatchEvent(
            new CustomEvent('corporations_list_result', {
                detail: { corporations },
            })
        );
    }

    private onBlackjackStateEvent(event: BlackjackStateEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const {
            state,
            dealerReveal,
            dealerCards,
            dealerTotal,
            playerHand,
            playerTotal,
            playerTurn,
            canDouble,
            canSplit,
            insuranceOffered,
            bet,
            messages,
            isDealer, // NEW
        } = parser.data ?? ({} as any);

        console.log('[🃏] Blackjack state received:', {
            state,
            dealerReveal,
            dealerCards,
            dealerTotal,
            playerHand,
            playerTotal,
            playerTurn,
            canDouble,
            canSplit,
            insuranceOffered,
            bet,
            messages,
            isDealer,
        });

        window.dispatchEvent(
            new CustomEvent('blackjack_state', {
                detail: {
                    state,
                    dealerReveal,
                    dealerCards,
                    dealerTotal,
                    playerHand,
                    playerTotal,
                    playerTurn,
                    canDouble,
                    canSplit,
                    insuranceOffered,
                    bet,
                    messages,
                    isDealer,
                },
            })
        );
    }

    // --- Blackjack MODULE bridge (unchanged) ---
    private onBlackjackModuleEvent(event: BlackjackModuleEvent): void {
        const parser = event.getParser();
        if (!parser) return;
        const hasInfo = parser.hasInfo;
        const minBet = parser.minBet;
        const maxBet = parser.maxBet;

        window.dispatchEvent(
            new CustomEvent('blackjack_module_result', {
                detail: { hasInfo, minBet, maxBet },
            })
        );
    }

    private onHighLowModuleEvent(event: HighLowModuleEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const hasInfo = parser.hasInfo;
        const minBet = parser.minBet;
        const maxBet = parser.maxBet;
        window.dispatchEvent(
            new CustomEvent('highlow_module_result', {
                detail: { hasInfo, minBet, maxBet },
            })
        );
    }

    private onSlotModuleEvent(event: SlotMachineModuleEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const hasInfo = parser.hasInfo;
        const minBet = parser.minBet; // expect 5
        const maxBet = parser.maxBet; // expect 25

        // Open your React UI (SlotView) the same way you do for High/Low:
        window.dispatchEvent(
            new CustomEvent('slotmachine_module_result', {
                detail: { hasInfo, minBet, maxBet },
            })
        );
    }
    private onWheelStateEvent = (event: WheelStateEvent): void => {
        const p = event.getParser();
        if (!p) return;

        window.dispatchEvent(
            new CustomEvent('bigwheel_state', {
                detail: {
                    state: p.state,
                    landedIndex: p.landedIndex,
                    landedFace: p.landedFace,
                    win: p.win,
                    betAmount: p.betAmount,
                    betFace: p.betFace,
                    faces: p.faces,
                },
            })
        );
    };

    private onSlotStateEvent(event: SlotMachineStateEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        // parser.state is already 'IDLE' | 'SPINNING' | 'ENDED'
        const stateStr = (parser.state || 'IDLE') as
            | 'IDLE'
            | 'SPINNING'
            | 'ENDED';

        const detail = {
            state: stateStr,
            reels: typeof parser.reels === 'number' ? parser.reels : 3,
            stops: Array.isArray(parser.stops) ? parser.stops : [],
            win: typeof parser.win === 'number' ? parser.win : 0,
            bet: typeof parser.bet === 'number' ? parser.bet : 0,
            messages: Array.isArray(parser.messages) ? parser.messages : [],
        };

        window.dispatchEvent(new CustomEvent('slotmachine_state', { detail }));
    }

    private onCorporationMembersEvent(event: CorporationMembersEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const members = parser.members.map((m) => {
            // build a friendly "x min/h/d ago" string from lastSeenMins
            let lastSeenAgo = '';
            const mins = m.lastSeenMins;

            if (mins <= 0) lastSeenAgo = 'Online now';
            else if (mins < 60) lastSeenAgo = `${mins} min ago`;
            else {
                const hours = Math.floor(mins / 60);
                if (hours < 24) lastSeenAgo = `${hours}h ago`;
                else {
                    const days = Math.floor(hours / 24);
                    lastSeenAgo = `${days}d ago`;
                }
            }

            return {
                userId: m.userId,
                username: m.username,
                figure: m.figure,
                rankId: m.rankId,
                rankName: m.rankName,
                rankOrder: m.rankOrder,
                pay: m.pay,

                // make sure these are always numbers
                weeklyShifts:
                    typeof m.weeklyShifts === 'number' ? m.weeklyShifts : 0,
                totalShifts:
                    typeof m.totalShifts === 'number' ? m.totalShifts : 0,

                lastSeenAgo,
            };
        });

        window.dispatchEvent(
            new CustomEvent('corporation_members_result', {
                detail: { members },
            })
        );
    }

    private onTimeOfDayEvent(event: TimeOfDayEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const data = parser.data; // { hhmm: "HH:mm", phase: "DAY" | "DUSK" | "NIGHT" | "DAWN" }

        window.dispatchEvent(
            new CustomEvent('time_of_day_update', {
                detail: data,
            })
        );
    }
    private onJukeboxStateEvent(event: JukeboxStateEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const data = parser.data; // { roomId, isOpen, isPlaying, currentVideoId, currentTitle, requestCost, expediteCost, queue[] }

        window.dispatchEvent(
            new CustomEvent('jukebox_state_update', {
                detail: data,
            })
        );
    }

    private onJukeboxOpenEvent(event: JukeboxOpenEvent): void {
        // No parser needed — open composer has no payload
        window.dispatchEvent(new CustomEvent('jukebox_open'));
    }

    // Jukebox Toast -> UI
    private onJukeboxToastEvent(event: JukeboxToastEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const data = parser.data; // { kind: 'info'|'success'|'error'|'warning', text: string }

        window.dispatchEvent(
            new CustomEvent('jukebox_toast', {
                detail: data,
            })
        );
    }

    private onVirtualRoomInfoEvent(event: VirtualRoomInfoEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const {
            virtualRoomId,
            name,
            type,
            credits,
            diamonds,
            inGameTime,
            phase,
            userCount, // 👈 grab it
        } = parser.data;

        // Main payload for the RightSideView, etc.
        window.dispatchEvent(
            new CustomEvent('virtual_room_info_update', {
                detail: {
                    virtualRoomId,
                    name,
                    type,
                    credits,
                    diamonds,
                    inGameTime,
                    phase,
                    userCount, // 👈 forward it
                },
            })
        );

        // Optional: keep balances in sync with any other listeners
        window.dispatchEvent(
            new CustomEvent('credit_balance_update', {
                detail: { amount: credits },
            })
        );
        window.dispatchEvent(
            new CustomEvent('diamond_balance_update', {
                detail: { amount: diamonds },
            })
        );
    }

    private onTurfCaptureProgressEvent(event: TurfCaptureProgressEvent): void {
        const p = event.getParser();
        if (!p) return;

        const d: any = p.data;

        const state = d.active
            ? d.contested
                ? 'contested'
                : 'capturing'
            : d.percent >= 100
            ? 'captured'
            : 'canceled';

        // pick colors from ANY of the names we might receive
        const primary =
            d.defenderPrimaryColor || d.ownerPrimaryColor || d.ownerClrA || '';
        const secondary =
            d.defenderSecondaryColor ||
            d.ownerSecondaryColor ||
            d.ownerClrB ||
            '';

        const ownerGangId = d.defenderGangId || 0;
        const ownerGangName = d.defenderGangName || '';

        const payload = {
            turfId: (d.turfId || '').toUpperCase(),
            virtualRoomId: d.virtualRoomId || 0,
            state,
            progress: Math.max(0, Math.min(100, Math.floor(d.percent || 0))),
            active: !!d.active,
            contested: !!d.contested,

            attackerGangId: d.attackerGangId || 0,
            attackerGangName: d.attackerGangName || '',

            // owner == defender during capture; after capture server sends attacker as defender
            defenderGangId: ownerGangId,
            defenderGangName: ownerGangName,

            // expose colors under ALL aliases GangClaimView understands
            defenderPrimaryColor: primary,
            defenderSecondaryColor: secondary,
            ownerPrimaryColor: primary,
            ownerSecondaryColor: secondary,
            ownerClrA: primary,
            ownerClrB: secondary,
        };

        // seed a tiny global cache so views can recover colors later if needed
        try {
            (window as any).__gangColorMap =
                (window as any).__gangColorMap || {};
            if (ownerGangId > 0 && primary && secondary) {
                (window as any).__gangColorMap[ownerGangId] = {
                    primary,
                    secondary,
                };
            }
        } catch {}

        window.dispatchEvent(
            new CustomEvent('turf_capture_progress', { detail: payload })
        );
        window.dispatchEvent(
            new CustomEvent('turf_capture_status', { detail: payload })
        );
    }

    onCorporationRanksEvent = (event: CorporationRanksEvent): void => {
        const parser = event.getParser();
        if (!parser) return;

        const corporationId = parser.corporationId;
        const ranks = parser.ranks;

        // Optional: log
        // console.log('[🏷️] Corporation ranks received:', corporationId, ranks);

        // Dispatch to React via DOM bridge
        window.dispatchEvent(
            new CustomEvent('corporation_ranks_result', {
                detail: { corporationId, ranks },
            })
        );
    };
    private onPaletteHexesEvent(event: PaletteHexesEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const data = {
            part: parser.part, // e.g. "cc"
            paletteId: parser.paletteId, // e.g. 2
            hexes: parser.hexes, // string[] of "#RRGGBB"
        };

        window.dispatchEvent(
            new CustomEvent('palette_hexes_result', { detail: data })
        );
    }

    private onWheelModuleEvent(event: WheelModuleEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        window.dispatchEvent(
            new CustomEvent('bigwheel_module_result', {
                detail: {
                    hasInfo: parser.hasInfo,
                    faces: parser.faces, // string[] (order = clockwise)
                    payouts: parser.payouts, // { "1":1, "2":2, ... }
                    minBet: parser.minBet,
                    maxBet: parser.maxBet,
                    spinMs: parser.spinMs,
                },
            })
        );
    }

    private onNameIconPurchaseOkEvent(event: NameIconPurchaseOkEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        window.dispatchEvent(
            new CustomEvent('name_icon_purchase_ok_raw', {
                detail: {
                    iconId: parser.iconId,
                    diamonds: parser.diamonds,
                },
            })
        );

        // Optional: keep wallet badge in sync like the bubbles flow
        window.dispatchEvent(
            new CustomEvent('diamond_balance_update', {
                detail: { amount: parser.diamonds },
            })
        );
    }
    private onCommandsListEvent(event: CommandsListMessageEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        // parser.commands is ICommandInfo[]
        const rawCommands = parser.commands;

        // Enrich with a default category so it matches CommandEntry
        const commands = rawCommands.map((cmd) => ({
            ...cmd,
            category: 'Generic',
        }));

        console.log('[📜 COMMANDS LIST] Received commands:', commands);

        // 1) Send data to React
        window.dispatchEvent(
            new CustomEvent('commands_list_update', {
                detail: commands,
            })
        );

        // 2) Open the window
        window.dispatchEvent(new Event('open_commands_list'));
    }

    private onOpenFishingModuleEvent(
        event: OpenFishingModuleMessageEvent
    ): void {
        const parser = event.getParser();
        if (!parser) return;

        // Dispatch raw payload (debug / optional use)
        window.dispatchEvent(
            new CustomEvent('open_fishing_raw', {
                detail: {
                    spotId: parser.spotId,
                    difficulty: parser.difficulty,
                    durationSeconds: parser.durationSeconds,
                    spotLabel: parser.spotLabel,
                },
            })
        );

        // Main event that your UI listens for
        window.dispatchEvent(
            new CustomEvent('open_fishing', {
                detail: {
                    spotId: parser.spotId,
                    difficulty: parser.difficulty,
                    durationSeconds: parser.durationSeconds,
                    spotLabel: parser.spotLabel,
                },
            })
        );
    }

    private onWantedListEvent(event: WantedListEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        // Parser already dispatches the update event automatically.
        // You can still log or perform extra actions if needed:
        const wantedUsers = parser.getWantedUsers();
        console.log('[🚨] Wanted list received:', wantedUsers);
    }

    private onHotelAlertEvent(event: HotelAlertEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const message = parser.message || '';

        //console.log('[🏨] Hotel alert received:', message);

        window.dispatchEvent(
            new CustomEvent('hotel_alert', {
                detail: { message },
            })
        );
    }

    private onDiscordVerificationRequiredEvent(event: DiscordVerificationRequiredEvent): void
{
const parser = event.getParser();
if(!parser) return;

const code = parser.code || '';
const serverTag = parser.serverTag || '';

window.dispatchEvent(new CustomEvent('discord_verification_required', {
detail: { code, serverTag }
}));
}

private onHaircutStylesOfferEvent(event: HaircutStylesOfferEvent): void
{
const parser = event.getParser();
if(!parser) return;

const offerId = parser.offerId || 0;
const stylistId = parser.stylistId || 0;
const stylistName = parser.stylistName || '';
const price = parser.price || 0;
const baseFigure = parser.baseFigure || '';

window.dispatchEvent(new CustomEvent('haircut_styles_offer', {
detail: {
offerId,
stylistId,
stylistName,
price,
baseFigure
}
}));
}

    private onOpenCorpJobBoardEvent(event: OpenCorpJobBoardEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const payload = {
            corpId: parser.corpId,
            corpName: parser.corpName,
            isManager: parser.isManager,
            openings: parser.openings,
            ranks: parser.ranks, // <-- NEW
        };

        window.dispatchEvent(
            new CustomEvent('corp_jobboard_open', { detail: payload })
        );
    }

    private onGuideStartEvent(event: QuickGuideEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        // Fire the overlay you already have wired in OnboardingOverlay.tsx
        window.dispatchEvent(new Event('start_quick_guide'));
    }

    private onHotRoomsListEvent(event: HotRoomsListMessageEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const rooms = parser.rooms;

        // Send data first
        window.dispatchEvent(
            new CustomEvent('hot_rooms_update', { detail: rooms })
        );

        // Then open the view
        window.dispatchEvent(new Event('open_hot_rooms'));
    }
    private onClickThroughStateEvent(event: ClickThroughStateEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const enabled = parser.enabled;

        // Global toggle (kept here too so it works even if parser stops doing it)
        (window as any).__ctEnabled = enabled;
        document.documentElement.classList.toggle('ct-on', enabled);
        console.log('clickthrough bridge sent');

        // Fire a UI bridge event for React/Vue/etc.
        window.dispatchEvent(
            new CustomEvent('click_through_state', { detail: { enabled } })
        );

        // (optional) toast/snackbar
        window.dispatchEvent(
            new CustomEvent('nitro_alert', {
                detail: enabled
                    ? 'Click-through enabled'
                    : 'Click-through disabled',
            })
        );
    }

    private onInventoryItemsEvent(event: InventoryMessageEvent): void {
        const parser = event.getParser();
        if (!parser) return;

        const items: InventoryContext[] = parser.items.map((item: any) => ({
            id: item.id,
            item_id: item.item_id,
            user_id: item.user_id,
            slot: item.slot,
            quantity: item.quantity,

            durability: item.durability,
            // ✅ correct key; optional fallback to legacy naming if present
            max_durability: item.max_durability ?? item.durability_max ?? 100,

            name: item.name,
            item_type: item.item_type,
            rarity: item.rarity,
            icon_path: item.icon_path,

            // keep extras if you use them; they won’t hurt
            last_used_at: item.last_used_at,
            description: item.description,
            cooldown_seconds: item.cooldown_seconds,
            damage_increase: item.damage_increase,
            health_increase: item.health_increase,
            max_stack: item.max_stack,
            created_at: item.created_at,
            stats: item.stats,
        }));

        console.log('Received Inventory:', items);

        // Make sure InventoryStore.setItems expects the InventoryContext shape with `max_durability`
        InventoryStore.setItems(items);

        window.dispatchEvent(new CustomEvent('inventory_update'));
    }

    private onOpponentStatsEvent(event: OpponentStatsEvent): void {
        const p = event.getParser();
        if (!p) return;

        const payload = {
            userId: p.userId,
            username: p.username,
            figure: p.figure,

            health: p.health,
            maxHealth: p.maxHealth,
            energy: p.energy,
            maxEnergy: p.maxEnergy,
            hunger: p.hunger,
            maxHunger: p.maxHunger,

            aggressionMs: p.aggressionMs,

            // xp / stats
            xp: p.xp,
            maxXP: p.maxXP,
            level: p.level,
            points: p.points,

            strength: p.strength,
            stamina: p.stamina,
            defense: p.defense,
            gathering: p.gatheringLevel,
            healthlevel: p.healthLevel,

            // fighting stats in canonical names
            punches: p.punchesThrown,
            damageGiven: p.damageDealt,
            damageReceived: p.damageReceived,
            kills: p.kills,
            deaths: p.deaths,

            shiftsWorked: p.shiftsWorked,
            arrests: p.arrests,

            isAggressive: p.isAggressive,

            // employment
            jobTitle: p.jobTitle,
            corporationName: p.corporationName,
            corporationIconUrl: p.corporationIconUrl,

            // gang
            gangId: p.gangId,
            gangName: p.gangName,
            gangIconKey: p.gangIconKey,
            gangPrimaryColor: p.gangPrimaryColor,
            gangSecondaryColor: p.gangSecondaryColor,
        };

        window.dispatchEvent(
            new CustomEvent('open-opponent-stats', { detail: payload })
        );
        window.dispatchEvent(
            new CustomEvent('user_inspect_stats', { detail: payload })
        );
        window.dispatchEvent(
            new CustomEvent('opponent_stats_update', { detail: payload })
        );
    }

    private onConnectionOpenedEvent(event: Event): void {
        this.logger.log('Connection Initialized');
    }

    private onConnectionClosedEvent(event: CloseEvent): void {
        this.logger.log('Connection Closed');
    }

    private onConnectionErrorEvent(event: Event): void {
        this.logger.log('Connection Error');
    }

    private onConnectionAuthenticatedEvent(event: NitroEvent): void {
        this.logger.log('Connection Authenticated');
        console.log('onConnectAuthenticatedEvent has been registered');

        if (this._connection) {
            this._connection.authenticated();
            this.registerMessageEvent(
                new ProfileAchievementsEvent(
                    this.onProfileAchievementsEvent.bind(this)
                )
            );

            this.registerMessageEvent(
                new TimeOfDayEvent(this.onTimeOfDayEvent.bind(this))
            );
            this.registerMessageEvent(
                new TurfCaptureProgressEvent(
                    this.onTurfCaptureProgressEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new HotRoomsListMessageEvent(
                    this.onHotRoomsListEvent.bind(this)
                )
            );

                 this.registerMessageEvent(
                new DiscordVerificationRequiredEvent(
                    this.onDiscordVerificationRequiredEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new DoorDashOrdersEvent(this.onDoorDashOrdersEvent.bind(this))
            );
            this.registerMessageEvent(
                new TaxiOpenListEvent(this.onTaxiOpenListEvent.bind(this))
            );
            this.registerMessageEvent(
                new TokensStoreEvent(this.onTokensStoreEvent.bind(this))
            );

            this.registerMessageEvent(
                new WheelStateEvent(this.onWheelStateEvent.bind(this))
            );

            this.registerMessageEvent(
                new HighLowModuleEvent(this.onHighLowModuleEvent.bind(this))
            );

            this.registerMessageEvent(
                new HighLowStateEvent(this.onHighLowStateEvent.bind(this))
            );

            this.registerMessageEvent(
                new GangRanksEvent(this.onGangRanksEvent.bind(this))
            );
             this.registerMessageEvent(
                new CasinoJackpotUpdateEvent(this.onCasinoJackpotUpdateEvent.bind(this))
            );
            // Now it's safe to register message event
            this.registerMessageEvent(
                new GangStatusEvent(this.onGangStatusEvent.bind(this))
            );
            // console.log("Registered gangstatusevent");
            this.registerMessageEvent(
                new InspectUserStatsEvent(
                    this.onInspectUserStatsEvent.bind(this)
                )
            );
               this.registerMessageEvent(
                new MythicalBlueprintOpenEvent(
                    this.onMythicalBlueprintOpenEvent.bind(this)
                )
            );
                this.registerMessageEvent(
                new MythicalBlueprintUpdateEvent(
                    this.onMythicalBlueprintUpdateEvent.bind(this)
                )
            );
                    this.registerMessageEvent(
                new MythicalBlueprintResultEvent(
                    this.onMythicalBlueprintResultEvent.bind(this)
                )
            );
            //console.log('Registered InspectUserStatsEvent');
            this.registerMessageEvent(
                new CorporationsListEvent(
                    this.onCorporationsListEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new MarketplaceModuleEvent(
                    this.onMarketplaceModuleEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new CorporationRanksEvent(
                    this.onCorporationRanksEvent.bind(this)
                )
            );

            // Trading

               this.registerMessageEvent(
                new TradeCloseEvent(
                    this.onTradeCloseEvent.bind(this)
                )
            );
              this.registerMessageEvent(
                new TradeOpenEvent(
                    this.onTradeOpenEvent.bind(this)
                )
            );
              this.registerMessageEvent(
                new TradeResultEvent(
                    this.onTradeResultEvent.bind(this)
                )
            );
              this.registerMessageEvent(
                new TradeUpdateEvent(
                    this.onTradeUpdateEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new WheelModuleEvent(this.onWheelModuleEvent.bind(this))
            );
            this.registerMessageEvent(
                new ItemOfferPopupEvent(this.onItemOfferPopupEvent.bind(this))
            );
            this.registerMessageEvent(
                new WantedListEvent(this.onWantedListEvent.bind(this))
            );

            this.registerMessageEvent(
                new UpdateWorkStatusEvent(
                    this.onUpdateWorkStatusEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new CorporationMembersEvent(
                    this.onCorporationMembersEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new GangMembersEvent(this.onGangMembersEvent.bind(this))
            );
            this.registerMessageEvent(
                new PoliceCallEvent(this.onPoliceCallEvent.bind(this))
            );
            this.registerMessageEvent(
                new HotelAlertEvent(this.onHotelAlertEvent.bind(this))
            );
            this.registerMessageEvent(
                new OpenCorpJobBoardEvent(
                    this.onOpenCorpJobBoardEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new SlotMachineStateEvent(this.onSlotStateEvent.bind(this))
            );
            this.registerMessageEvent(
                new SlotMachineModuleEvent(this.onSlotModuleEvent.bind(this))
            );
            this.registerMessageEvent(
                new PaletteHexesEvent(this.onPaletteHexesEvent.bind(this))
            );
            this.registerMessageEvent(
                new EquippedLoadoutEvent(this.onEquippedLoadoutEvent.bind(this))
            );
            this.registerMessageEvent(
                new VirtualRoomInfoEvent(this.onVirtualRoomInfoEvent.bind(this))
            );
            this.registerMessageEvent(new RoleplayCombatCooldownEvent(this.onRoleplayCombatCooldownEvent.bind(this)));
            this.registerMessageEvent(
                new OpenSellModuleEvent(this.onOpenSellModuleEvent.bind(this))
            );
            this.registerMessageEvent(
                new CancelGatherEvent(this.onCancelGatherEvent.bind(this))
            );
            this.registerMessageEvent(
                new GatherProgressEvent(this.onGatherProgressEvent.bind(this))
            );
            this.registerMessageEvent(
                new InventoryMessageEvent(this.onInventoryItemsEvent.bind(this))
            );
            this.registerMessageEvent(
                new GangInviteEvent(this.onGangInviteEvent.bind(this))
            );
            this.registerMessageEvent(
                new OpponentStatsEvent(this.onOpponentStatsEvent.bind(this))
            );
            this.registerMessageEvent(
                new ClickThroughStateEvent(
                    this.onClickThroughStateEvent.bind(this)
                )
            );
                this.registerMessageEvent(
                new NameIconEquippedEvent(
                    this.onNameIconEquippedEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new HaircutStylesOfferEvent(
                    this.onHaircutStylesOfferEvent.bind(this)
                )
            );

            //console.log('Registered InventoryMessageEvent');

            this.registerMessageEvent(
                new UserStatsEvent(this.onUserStatsEvent.bind(this))
            );
            this.registerMessageEvent(
                new CommandsListMessageEvent(
                    this.onCommandsListEvent.bind(this)
                )
            );
            //console.log('Registered UserStatsEvent');
            this.registerMessageEvent(
                new ZeusHealthEvent(this.onZeusStatsEvent.bind(this))
            );
            //console.log('Registered ZeusStatsEvent');
            this.registerMessageEvent(
                new UserLiveFeedEvent(this.onUserLiveFeedEvent.bind(this))
            );
               this.registerMessageEvent(
                new ArenaLeaderboardEvent(this.onArenaLeaderboardEvent.bind(this))
            );
            this.registerMessageEvent(
                new ATMOpenMessageEvent(this.onATMOpenEvent.bind(this))
            );

            this.registerMessageEvent(
                new BlackjackModuleEvent(this.onBlackjackModuleEvent.bind(this))
            );

            this.registerMessageEvent(
                new BlackjackStateEvent(this.onBlackjackStateEvent.bind(this))
            );
            this.registerMessageEvent(
                new QuickGuideEvent(this.onGuideStartEvent.bind(this))
            );
                 this.registerMessageEvent(
                new HeistTimerEvent(this.onHeistTimerEvent.bind(this))
            );
                 this.registerMessageEvent(
                new HeistTimerStopEvent(this.onHeistTimerStopEvent.bind(this))
            );
            this.registerMessageEvent(
                new JukeboxStateEvent(this.onJukeboxStateEvent.bind(this))
            );
            this.registerMessageEvent(
                new JukeboxOpenEvent(this.onJukeboxOpenEvent.bind(this))
            );

            this.registerMessageEvent(
                new JukeboxToastEvent(this.onJukeboxToastEvent.bind(this))
            );
            this.registerMessageEvent(
                new ChatBubbleShopEvent(this.onChatBubbleShopEvent.bind(this))
            );
            this.registerMessageEvent(
                new OpenAvatarEditorEvent(
                    this.onOpenAvatarEditorEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new ChatBubblePurchaseOkEvent(
                    this.onChatBubblePurchaseOkEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new NameIconShopEvent(this.onNameIconShopEvent.bind(this))
            );
            this.registerMessageEvent(
                new NameIconPurchaseOkEvent(
                    this.onNameIconPurchaseOkEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new OpenFishingModuleMessageEvent(
                    this.onOpenFishingModuleEvent.bind(this)
                )
            );
            this.registerMessageEvent(
                new ArenaStatusMessageEvent(this.onArenaStatusEvent.bind(this))
            );
        }
    }

    public connectionInit(socketUrl: string): void {
        this.logger.log(`Initializing Connection: ${socketUrl}`);
    }

    public registerMessageEvent(event: IMessageEvent): IMessageEvent {
        if (this._connection) this._connection.addMessageEvent(event);
        //console.log("Recieved header:", event)

        return event;
    }

    public removeMessageEvent(event: IMessageEvent): void {
        if (!this._connection) return;

        this._connection.removeMessageEvent(event);
    }

    public get demo(): NitroCommunicationDemo {
        return this._demo;
    }

    public get connection(): IConnection {
        return this._connection;
    }
}
