// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/ArenaStatusParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export class ArenaStatusParser implements IMessageParser {
    private _inQueue: boolean = false;
    private _isMatched: boolean = false;

    private _selfId: number = 0;
    private _selfUsername: string = '';
    private _selfFigure: string = '';
    private _selfLevel: number = 0;
    private _selfWins: number = 0;
    private _selfLosses: number = 0;

    private _oppId: number = 0;
    private _oppUsername: string = '';
    private _oppFigure: string = '';
    private _oppLevel: number = 0;
    private _oppWins: number = 0;
    private _oppLosses: number = 0;

    public flush(): boolean {
        this._inQueue = false;
        this._isMatched = false;

        this._selfId = 0;
        this._selfUsername = '';
        this._selfFigure = '';
        this._selfLevel = 0;
        this._selfWins = 0;
        this._selfLosses = 0;

        this._oppId = 0;
        this._oppUsername = '';
        this._oppFigure = '';
        this._oppLevel = 0;
        this._oppWins = 0;
        this._oppLosses = 0;

        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        // MUST match ArenaStatusComposer.java exactly:
        // bool inQueue
        // bool isMatched
        // int selfId
        // string selfUsername
        // string selfFigure
        // int selfLevel
        // int selfWins
        // int selfLosses
        // int oppId
        // string oppUsername
        // string oppFigure
        // int oppLevel
        // int oppWins
        // int oppLosses

        this._inQueue = wrapper.readBoolean();
        this._isMatched = wrapper.readBoolean();

        this._selfId = wrapper.readInt();
        this._selfUsername = wrapper.readString();
        this._selfFigure = wrapper.readString();
        this._selfLevel = wrapper.readInt();
        this._selfWins = wrapper.readInt();
        this._selfLosses = wrapper.readInt();

        this._oppId = wrapper.readInt();
        this._oppUsername = wrapper.readString();
        this._oppFigure = wrapper.readString();
        this._oppLevel = wrapper.readInt();
        this._oppWins = wrapper.readInt();
        this._oppLosses = wrapper.readInt();

        return true;
    }

    // ----- getters -----

    public get inQueue(): boolean {
        return this._inQueue;
    }

    public get isMatched(): boolean {
        return this._isMatched;
    }

    public get selfId(): number {
        return this._selfId;
    }

    public get selfUsername(): string {
        return this._selfUsername;
    }

    public get selfFigure(): string {
        return this._selfFigure;
    }

    public get selfLevel(): number {
        return this._selfLevel;
    }

    public get selfWins(): number {
        return this._selfWins;
    }

    public get selfLosses(): number {
        return this._selfLosses;
    }

    public get oppId(): number {
        return this._oppId;
    }

    public get oppUsername(): string {
        return this._oppUsername;
    }

    public get oppFigure(): string {
        return this._oppFigure;
    }

    public get oppLevel(): number {
        return this._oppLevel;
    }

    public get oppWins(): number {
        return this._oppWins;
    }

    public get oppLosses(): number {
        return this._oppLosses;
    }
}
