// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/BlackJackModuleParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export class BlackjackModuleParser implements IMessageParser {
    private _hasInfo: boolean = false;
    private _tableId: number | null = null;
    private _minBet: number | null = null;
    private _maxBet: number | null = null;

    public flush(): boolean {
        this._hasInfo = false;
        this._tableId = null;
        this._minBet = null;
        this._maxBet = null;
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        this._hasInfo = wrapper.readBoolean();

        if (this._hasInfo) {
            this._tableId = wrapper.readInt();
            this._minBet = wrapper.readInt();
            this._maxBet = wrapper.readInt();
        }

        return true;
    }

    public get hasInfo(): boolean {
        return this._hasInfo;
    }
    public get tableId(): number | null {
        return this._tableId;
    }
    public get minBet(): number | null {
        return this._minBet;
    }
    public get maxBet(): number | null {
        return this._maxBet;
    }
}
