// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/BulkFurnitureFloorAddParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export interface IBulkAddedFloorItem {
    itemId: number;
    spriteId: number;
    x: number;
    y: number;
    rot: number;
    z: string;
    heightOrEmpty: string;

    extraInt: number; // gift ribbon code / songId / 1
    extraData: string; // ALWAYS string in this matching version

    groupId: number;
    isUsable: number;
    userId: number;
    ownerName: string;
}

export class BulkAddFloorItemsParser implements IMessageParser {
    private _items: IBulkAddedFloorItem[] = [];

    public flush(): boolean {
        this._items = [];
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        const count = wrapper.readInt();
        this._items = new Array<IBulkAddedFloorItem>(count);

        for (let i = 0; i < count; i++) {
            const itemId = wrapper.readInt();
            const spriteId = wrapper.readInt();
            const x = wrapper.readInt();
            const y = wrapper.readInt();
            const rot = wrapper.readInt();
            const z = wrapper.readString();
            const heightOrEmpty = wrapper.readString();

            const extraInt = wrapper.readInt();
            const extraData = wrapper.readString(); // <-- string by contract

            const groupId = wrapper.readInt();
            const isUsable = wrapper.readInt();
            const userId = wrapper.readInt();
            const ownerName = wrapper.readString();

            this._items[i] = {
                itemId,
                spriteId,
                x,
                y,
                rot,
                z,
                heightOrEmpty,
                extraInt,
                extraData,
                groupId,
                isUsable,
                userId,
                ownerName,
            };
        }

        return true;
    }

    public get items(): IBulkAddedFloorItem[] {
        return this._items;
    }
}
