// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/BulkFurnitureFloorRemoveParser.ts

import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer/src/core';

export interface IBulkRemovedFloorItem {
    itemId: number; // from string in the wire format
    isExpired: boolean; // mirrors your single remove boolean
    userId: number; // 0 when noUser=true on server
    delay: number; // trailing int (kept for parity)
}

export class BulkFurnitureFloorRemoveParser implements IMessageParser {
    private _items: IBulkRemovedFloorItem[] = [];

    public flush(): boolean {
        this._items = [];
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        const count = wrapper.readInt();
        this._items = new Array<IBulkRemovedFloorItem>(count);

        for (let i = 0; i < count; i++) {
            const itemIdStr = wrapper.readString();
            const isExpired = wrapper.readBoolean();
            const userId = wrapper.readInt();
            const delay = wrapper.readInt();

            this._items[i] = {
                itemId: parseInt(itemIdStr, 10) || 0,
                isExpired,
                userId,
                delay,
            };
        }

        return true;
    }

    public get items(): IBulkRemovedFloorItem[] {
        return this._items;
    }
}
