// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/ChatBubblePurchaseOkParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export class ChatBubblePurchaseOkParser implements IMessageParser {
    private _bubbleId = 0;
    private _diamonds = 0;

    public flush(): boolean {
        this._bubbleId = 0;
        this._diamonds = 0;
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        // must match server order:
        // this.response.appendInt(bubbleId);
        // this.response.appendInt(diamonds);
        this._bubbleId = wrapper.readInt();
        this._diamonds = wrapper.readInt();

        // Broadcast to the React layer (what DiamondsStoreView expects)
        window.dispatchEvent(
            new CustomEvent('chat_bubble_purchase_ok_raw', {
                detail: {
                    bubbleId: this._bubbleId,
                    diamonds: this._diamonds,
                },
            })
        );

        return true;
    }

    public get bubbleId(): number {
        return this._bubbleId;
    }
    public get diamonds(): number {
        return this._diamonds;
    }
}
