// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/ChatBubbleShopParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export interface ChatBubbleShopItem {
    id: number; // bubbleId
    price: number; // diamonds
    owned: boolean;
    equipped: boolean;
}

export class ChatBubbleShopParser implements IMessageParser {
    private _items: ChatBubbleShopItem[] = [];
    private _equippedId = 0;
    private _diamonds?: number; // optional; wire if you later send balance

    public flush(): boolean {
        this._items = [];
        this._equippedId = 0;
        this._diamonds = undefined;
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        const count = wrapper.readInt();
        const items: ChatBubbleShopItem[] = [];

        for (let i = 0; i < count; i++) {
            const id = wrapper.readInt();
            const price = wrapper.readInt();
            const owned = wrapper.readBoolean();
            const equipped = wrapper.readBoolean();
            items.push({ id, price, owned, equipped });
        }

        this._items = items;

        // derive equipped id if present
        const eq = items.find((i) => i.equipped);
        this._equippedId = eq ? eq.id : 0;

        // 🔔 Dispatch the shape SettingsView expects:
        // detail: { items: [{ bubbleId, price, owned, equipped }], diamonds?, equippedId? }
        window.dispatchEvent(
            new CustomEvent('chat_bubble_shop', {
                detail: {
                    items: items.map((i) => ({
                        bubbleId: i.id,
                        price: i.price,
                        owned: i.owned,
                        equipped: i.equipped,
                    })),
                    diamonds: this._diamonds, // ok to be undefined
                    equippedId: this._equippedId,
                },
            })
        );

        return true;
    }

    public get items(): ChatBubbleShopItem[] {
        return this._items;
    }
    public get equippedId(): number {
        return this._equippedId;
    }
    public get diamonds(): number | undefined {
        return this._diamonds;
    }
}
