// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/CorporationsMembersParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export interface CorporationMemberData {
    userId: number;
    username: string;
    figure: string;
    rankId: number;
    rankName: string;
    rankOrder: number;
    pay: number;
    weeklyShifts: number;
    totalShifts: number;
    lastSeenMins: number;
}

export class CorporationMembersParser implements IMessageParser {
    private _members: CorporationMemberData[] = [];

    public flush(): boolean {
        this._members = [];
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        const total = wrapper.readInt();
        const members: CorporationMemberData[] = new Array(total);

        for (let i = 0; i < total; i++) {
            members[i] = {
                userId: wrapper.readInt(),
                username: wrapper.readString(),
                figure: wrapper.readString(),
                rankId: wrapper.readInt(),
                rankName: wrapper.readString(),
                rankOrder: wrapper.readInt(),
                pay: wrapper.readInt(),
                weeklyShifts: wrapper.readInt(),
                totalShifts: wrapper.readInt(),
                lastSeenMins: wrapper.readInt(),
            };
        }

        this._members = members;
        return true;
    }

    public get members(): CorporationMemberData[] {
        return this._members;
    }
}
