// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/CorporationRankParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export interface CorporationRankData {
    rankId: number;
    rankName: string;
    rankOrder: number; // higher number = higher rank
    pay: number;
    uniform: string; // work_figure
    isPolice: boolean;
    isParamedic: boolean;
    isSupervisor: boolean;
    isManager: boolean;
    isHospital: boolean;
    isArmory: boolean;
}

export class CorporationRanksParser implements IMessageParser {
    private _corporationId: number = -1;
    private _ranks: CorporationRankData[] = [];

    public flush(): boolean {
        this._corporationId = -1;
        this._ranks = [];
        return true;
    }

    /**
     * Field order (must match server composer):
     * 1) corporationId (int)
     * 2) count (int)
     * For each rank:
     * 3) rankId (int)
     * 4) rankName (string)
     * 5) rankOrder (int)
     * 6) pay (int)
     * 7) uniform (string)
     * 8) isPolice (bool)
     * 9) isParamedic (bool)
     * 10) isSupervisor (bool)
     * 11) isManager (bool)
     * 12) isHospital (bool)
     * 13) isArmory (bool)
     */
    public parse(wrapper: IMessageDataWrapper): boolean {
        this._corporationId = wrapper.readInt();

        const total = wrapper.readInt();
        const list: CorporationRankData[] = new Array(total);

        for (let i = 0; i < total; i++) {
            const rankId = wrapper.readInt();
            const rankName = wrapper.readString();
            const rankOrder = wrapper.readInt();
            const pay = wrapper.readInt();
            const uniform = wrapper.readString();
            const isPolice = wrapper.readBoolean();
            const isParamedic = wrapper.readBoolean();
            const isSupervisor = wrapper.readBoolean();
            const isManager = wrapper.readBoolean();
            const isHospital = wrapper.readBoolean();
            const isArmory = wrapper.readBoolean();

            list[i] = {
                rankId,
                rankName,
                rankOrder,
                pay,
                uniform,
                isPolice,
                isParamedic,
                isSupervisor,
                isManager,
                isHospital,
                isArmory,
            };
        }

        this._ranks = list;
        return true;
    }

    // Getters
    public get corporationId(): number {
        return this._corporationId;
    }
    public get ranks(): CorporationRankData[] {
        return this._ranks;
    }

    // Optional helpers
    public get hasData(): boolean {
        return this._corporationId > -1;
    }
    public get ranksByOrderDesc(): CorporationRankData[] {
        return [...this._ranks].sort((a, b) => b.rankOrder - a.rankOrder);
    }
}
