// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/DoorDashOrdersParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export interface DoorDashOrderData {
    orderId: number;
    username: string;
    virtualRoomId: number;
    virtualRoomName: string;
    ageSeconds: number; // age in seconds (already computed by composer)
}

export class DoorDashOrdersParser implements IMessageParser {
    private _orders: DoorDashOrderData[] = [];

    public flush(): boolean {
        this._orders = [];
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        this._orders = [];

        // MUST match DoorDashOrdersComposer exactly:
        // count:int,
        // orderId:int, username:string, vId:int, vName:string, ageSeconds:int
        const count = wrapper.readInt();

        for (let i = 0; i < count; i++) {
            const orderId = wrapper.readInt();
            const username = wrapper.readString();
            const virtualRoomId = wrapper.readInt();
            const virtualRoomName = wrapper.readString();
            const ageSeconds = wrapper.readInt();

            this._orders.push({
                orderId,
                username,
                virtualRoomId,
                virtualRoomName,
                ageSeconds,
            });
        }

        return true;
    }

    public get orders(): DoorDashOrderData[] {
        return this._orders;
    }
}
