// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/EquippedLoadoutParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export interface WeaponData {
    userId: number;
    itemId: number;
    name: string;
    damage: number;
    icon: string;
}

export interface ShieldData {
    userId: number;
    itemId: number;
    name: string;
    defense: number;
    icon: string;
}

export interface PotionData {
    userId: number;
    itemId: number;
    name: string;
    type: string; // "health" | "energy" | "hunger" | "rainbow" | etc.
    remaining: number; // seconds or ticks
    icon: string;
}

export class EquippedLoadoutParser implements IMessageParser {
    private _weapon: WeaponData | null = null;
    private _shield: ShieldData | null = null;
    private _potion: PotionData | null = null;

    public flush(): boolean {
        this._weapon = null;
        this._shield = null;
        this._potion = null;
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        // Weapon
        const hasWeapon = wrapper.readBoolean();
        if (hasWeapon) {
            this._weapon = {
                userId: wrapper.readInt(),
                itemId: wrapper.readInt(),
                name: wrapper.readString(),
                damage: wrapper.readInt(),
                icon: wrapper.readString(),
            };
        } else {
            this._weapon = null;
        }

        // Shield
        const hasShield = wrapper.readBoolean();
        if (hasShield) {
            this._shield = {
                userId: wrapper.readInt(),
                itemId: wrapper.readInt(),
                name: wrapper.readString(),
                defense: wrapper.readInt(),
                icon: wrapper.readString(),
            };
        } else {
            this._shield = null;
        }

        // Potion
        const hasPotion = wrapper.readBoolean();
        if (hasPotion) {
            this._potion = {
                userId: wrapper.readInt(),
                itemId: wrapper.readInt(),
                name: wrapper.readString(),
                type: wrapper.readString(),
                remaining: wrapper.readInt(),
                icon: wrapper.readString(),
            };
        } else {
            this._potion = null;
        }

        return true;
    }

    public get weapon(): WeaponData | null {
        return this._weapon;
    }
    public get shield(): ShieldData | null {
        return this._shield;
    }
    public get potion(): PotionData | null {
        return this._potion;
    }
}
