// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/GangRanksParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export interface GangRankData {
    id: number;
    name: string;
    position: number;
    canInvite: boolean;
    canKick: boolean;
    canAlert: boolean;
    canEditRanks: boolean;
    canAccessBank: boolean;
    administrator: boolean;
}

export class GangRanksParser implements IMessageParser {
    private _ranks: GangRankData[] = [];

    public flush(): boolean {
        this._ranks = [];
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        const count = wrapper.readInt();

        for (let i = 0; i < count; i++) {
            this._ranks.push({
                id: wrapper.readInt(), // id
                name: wrapper.readString(), // name
                position: wrapper.readInt(), // position
                canInvite: wrapper.readBoolean(), // caninvite
                canKick: wrapper.readBoolean(),
                canAlert: wrapper.readBoolean(),
                canEditRanks: wrapper.readBoolean(),
                canAccessBank: wrapper.readBoolean(),
                administrator: wrapper.readBoolean(),
            });
        }
    console.log('[Parser] -> GangRanksParser.ts recognized.');
        return true;
    }

    public get ranks(): GangRankData[] {
        return this._ranks;
    }
}
