// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/GangMembersParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export interface GangMemberWire {
    userId: number;
    username: string;
    rankId: number;
    rankName: string;
    figure: string;
}

export class GangMembersParser implements IMessageParser {
    private _members: GangMemberWire[] = [];

    public flush(): boolean {
        this._members = [];
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;
        const count = wrapper.readInt();
        const list: GangMemberWire[] = new Array(count);
        for (let i = 0; i < count; i++) {
            list[i] = {
                userId: wrapper.readInt(),
                username: wrapper.readString(),
                rankId: wrapper.readInt(),
                rankName: wrapper.readString(),
                figure: wrapper.readString(),
            };
        }
        this._members = list;

        // notify React bridge
        window.dispatchEvent(
            new CustomEvent('gang_members_result', {
                detail: { members: list },
            })
        );
        // Optional alias
        window.dispatchEvent(
            new CustomEvent('gang_members_received', {
                detail: { members: list },
            })
        );

        return true;
    }

    public get members(): GangMemberWire[] {
        return this._members;
    }
}
