// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/GangStatusParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export interface GangStatusData {
    inGang: boolean;
    gangName: string;
    gangRank: number;
    primaryColor: string;
    secondaryColor: string;
    xp: number;
    maxXp: number;
    kills: number;
    deaths: number;
    robberies: number;
    pickpockets: number;
    totalDamage: number;
    turfsControlled: number;
    isGangAdmin: boolean;
    iconKey: string;
    turfsOwned: string[]; // NEW
}

export class GangStatusParser implements IMessageParser {
    private _data: GangStatusData = this.createEmptyData();

    private createEmptyData(): GangStatusData {
        return {
            inGang: false,
            gangName: '',
            gangRank: 0,
            primaryColor: '',
            secondaryColor: '',
            xp: 0,
            maxXp: 0,
            kills: 0,
            deaths: 0,
            robberies: 0,
            pickpockets: 0,
            totalDamage: 0,
            turfsControlled: 0,
            isGangAdmin: false,
            iconKey: '',
            turfsOwned: [],
        };
    }

    public flush(): boolean {
        this._data = this.createEmptyData();
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        // Order MUST match server composer
        this._data.inGang = wrapper.readBoolean();
        this._data.gangName = wrapper.readString();
        this._data.gangRank = wrapper.readInt();
        this._data.primaryColor = wrapper.readString();
        this._data.secondaryColor = wrapper.readString();

        this._data.xp = wrapper.readInt();
        this._data.maxXp = wrapper.readInt();
        this._data.kills = wrapper.readInt();
        this._data.deaths = wrapper.readInt();
        this._data.robberies = wrapper.readInt();
        this._data.pickpockets = wrapper.readInt();
        this._data.totalDamage = wrapper.readInt();
        this._data.turfsControlled = wrapper.readInt();
        this._data.isGangAdmin = wrapper.readBoolean();

        // iconKey (safe read for backward compatibility)
        try {
            // optional safety check for remaining bytes
            // @ts-ignore
            if (
                typeof (wrapper as any).bytesAvailable === 'number' &&
                (wrapper as any).bytesAvailable <= 0
            ) {
                this._data.iconKey = '';
            } else {
                this._data.iconKey = wrapper.readString() || '';
            }
        } catch {
            this._data.iconKey = '';
        }

        // --- NEW: Read turf bag (count + key/boolean pairs)
        try {
            const tCount = wrapper.readInt();
            const owned: string[] = [];

            for (let i = 0; i < tCount; i++) {
                const key = wrapper.readString();
                const has = wrapper.readBoolean();
                if (has) owned.push(key);
            }

            this._data.turfsOwned = owned;
        } catch {
            // old servers may not send this yet
            this._data.turfsOwned = [];
        }

        // Normalize color format
        const withHash = (hex: string) =>
            hex ? (hex.startsWith('#') ? hex : `#${hex}`) : '';
        this._data.primaryColor = withHash(this._data.primaryColor);
        this._data.secondaryColor = withHash(this._data.secondaryColor);

        // Dispatch to UI bridge
        window.dispatchEvent(
            new CustomEvent('gang_status_result', { detail: this._data })
        );
        window.dispatchEvent(
            new CustomEvent('gang_status_received', { detail: this._data })
        );

        return true;
    }

    public get data(): GangStatusData {
        return this._data;
    }
}
