// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/HighLowStateParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export interface HighLowMessage {
    text: string;
    level: string; // "info" | "warn" | "error"
}

export class HighLowStateParser implements IMessageParser {
    private _state: string = 'IDLE';
    private _turn: string = 'PLAYER';
    private _currentCard: string = '?';
    private _canAct: boolean = false;
    private _bet: number = 0;
    private _messages: HighLowMessage[] = [];

    flush(): boolean {
        this._state = 'IDLE';
        this._turn = 'PLAYER';
        this._currentCard = '?';
        this._canAct = false;
        this._bet = 0;
        this._messages = [];
        return true;
    }

    parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        // MUST match server compose order:
        // state, turn, currentCard, canAct, bet, messages[]
        this._state = wrapper.readString();
        this._turn = wrapper.readString();
        this._currentCard = wrapper.readString();
        this._canAct = wrapper.readBoolean();
        this._bet = wrapper.readInt();

        const count = wrapper.readInt();
        this._messages = [];
        for (let i = 0; i < count; i++) {
            const text = wrapper.readString();
            const level = wrapper.readString();
            this._messages.push({ text, level });
        }

        return true;
    }

    public get state(): string {
        return this._state;
    }
    public get turn(): string {
        return this._turn;
    }
    public get currentCard(): string {
        return this._currentCard;
    }
    public get canAct(): boolean {
        return this._canAct;
    }
    public get bet(): number {
        return this._bet;
    }
    public get messages(): HighLowMessage[] {
        return this._messages;
    }
}
