// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/JukeboxStateParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

/** One queued track in the jukebox list */
export interface JukeboxQueueEntry {
    id: number;
    videoId: string;
    title: string;
    requestedBy: string;
    expedited: boolean;
}

/** Full jukebox snapshot */
export interface JukeboxStateData {
    roomId: number;
    isOpen: boolean;
    isPlaying: boolean;
    currentVideoId: string;
    currentTitle: string;
    requestCost: number;
    expediteCost: number;
    startedAt: number; // NEW: unix seconds (0 if idle)
    queue: JukeboxQueueEntry[];
}

export class JukeboxStateParser implements IMessageParser {
    private _data: JukeboxStateData = {
        roomId: 0,
        isOpen: false,
        isPlaying: false,
        currentVideoId: '',
        currentTitle: '',
        requestCost: 0,
        expediteCost: 0,
        startedAt: 0,
        queue: [],
    };

    public flush(): boolean {
        this._data = {
            roomId: 0,
            isOpen: false,
            isPlaying: false,
            currentVideoId: '',
            currentTitle: '',
            requestCost: 0,
            expediteCost: 0,
            startedAt: 0,
            queue: [],
        };
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        /**
         * MUST match server JukeboxStateComposer order:
         * roomId:int
         * isOpen:bool
         * isPlaying:bool
         * currentVideoId:string
         * currentTitle:string
         * requestCost:int
         * expediteCost:int
         * startedAt:int // NEW
         * queueCount:int
         * loop queueCount times:
         * id:int
         * videoId:string
         * title:string
         * requestedBy:string
         * expedited:bool
         */
        const roomId = wrapper.readInt();
        const isOpen = wrapper.readBoolean();
        const isPlaying = wrapper.readBoolean();
        const curVid = wrapper.readString();
        const curTitle = wrapper.readString();
        const requestCost = wrapper.readInt();
        const expediteCost = wrapper.readInt();
        const startedAt = wrapper.readInt(); // NEW

        const count = wrapper.readInt();
        const queue: JukeboxQueueEntry[] = new Array(count);
        for (let i = 0; i < count; i++) {
            queue[i] = {
                id: wrapper.readInt(),
                videoId: wrapper.readString(),
                title: wrapper.readString(),
                requestedBy: wrapper.readString(),
                expedited: wrapper.readBoolean(),
            };
        }

        this._data = {
            roomId,
            isOpen,
            isPlaying,
            currentVideoId: curVid,
            currentTitle: curTitle,
            requestCost,
            expediteCost,
            startedAt,
            queue,
        };

        return true;
    }

    public get data(): JukeboxStateData {
        return this._data;
    }
}
