// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/JukeboxToastParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export type JukeboxToastKind = 'info' | 'success' | 'error' | 'warning';

export interface JukeboxToastData {
    kind: JukeboxToastKind; // semantic type for UI styling
    text: string; // message to show
    // Optional: server may include extra fields later, e.g., seconds:int
}

export class JukeboxToastParser implements IMessageParser {
    private _data: JukeboxToastData = { kind: 'info', text: '' };

    public flush(): boolean {
        this._data = { kind: 'info', text: '' };
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        /**
         * MUST match server JukeboxToastComposer order:
         * kind:string // "info" | "success" | "error" | "warning"
         * text:string
         */
        const kind = wrapper.readString() as JukeboxToastKind;
        const text = wrapper.readString();

        this._data = { kind, text };
        return true;
    }

    public get data(): JukeboxToastData {
        return this._data;
    }
}
