// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/MarketplaceModuleParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

/* =========================
DTOs
========================= */
export interface MarketplaceListingDTO {
    listingId: number;
    itemId: number;
    baseId: number;

    name: string;
    description: string;

    price: number;
    quantity: number;

    durabilityCurrent: number;
    durabilityMax: number;

    sellerName: string;
    sellerIsYou: boolean;

    secondsRemaining: number;

    categoryId: number;
    isTradable: boolean;
}

export interface MarketplaceModuleData {
    defaultCategoryId: number;
    prefillQuery: string;
    canSell: boolean;
    userCredits: number;

    page: number;
    pageSize: number;
    totalResults: number;

    listings: MarketplaceListingDTO[];
}

/* =========================
Parser
========================= */
export class MarketplaceModuleParser implements IMessageParser {
    private _data: MarketplaceModuleData = this.createEmpty();

    private createEmpty(): MarketplaceModuleData {
        return {
            defaultCategoryId: 0,
            prefillQuery: '',
            canSell: false,
            userCredits: 0,
            page: 1,
            pageSize: 25,
            totalResults: 0,
            listings: [],
        };
    }

    public flush(): boolean {
        this._data = this.createEmpty();
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        try {
            // ---- Header ----
            const defaultCategoryId = wrapper.readInt();
            const prefillQuery = wrapper.readString();
            const canSell = wrapper.readBoolean();
            const userCredits = wrapper.readInt();

            const page = wrapper.readInt();
            const pageSize = wrapper.readInt();
            const totalResults = wrapper.readInt();

            // ---- Listings ----
            const count = wrapper.readInt();
            const listings: MarketplaceListingDTO[] = [];

            for (let i = 0; i < count; i++) {
                const listingId = wrapper.readInt();
                const itemId = wrapper.readInt();
                const baseId = wrapper.readInt();

                const name = wrapper.readString();
                const description = wrapper.readString();

                const price = wrapper.readInt();
                const quantity = wrapper.readInt();

                const durabilityCurrent = wrapper.readInt();
                const durabilityMax = wrapper.readInt();

                const sellerName = wrapper.readString();
                const sellerIsYou = wrapper.readBoolean();

                const secondsRemaining = wrapper.readInt();

                const categoryId = wrapper.readInt();
                const isTradable = wrapper.readBoolean();

                listings.push({
                    listingId,
                    itemId,
                    baseId,
                    name,
                    description,
                    price,
                    quantity,
                    durabilityCurrent,
                    durabilityMax,
                    sellerName,
                    sellerIsYou,
                    secondsRemaining,
                    categoryId,
                    isTradable,
                });
            }

            this._data = {
                defaultCategoryId,
                prefillQuery,
                canSell,
                userCredits,
                page,
                pageSize,
                totalResults,
                listings,
            };

            // Fire event for UI
            window.dispatchEvent(
                new CustomEvent<MarketplaceModuleData>(
                    'marketplace_module_result',
                    {
                        detail: this._data,
                    }
                )
            );

            return true;
        } catch (e) {
            console.error('[MarketplaceModuleParser] Parse failed:', e);
            return false;
        }
    }

    public get data(): MarketplaceModuleData {
        return this._data;
    }
}
