// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/OpenCorpJobBoardParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export interface ICorpJobOpeningData {
    id: number;
    title: string;
    openSlots: number;
    isOpen: boolean;
    createdByUserId: number;
    createdAt: number; // unix seconds
}

export interface ICorpRankOption {
    id: number;
    name: string;
}

export class OpenCorpJobBoardParser implements IMessageParser {
    private _corpId: number = 0;
    private _corpName: string = '';
    private _isManager: boolean = false;
    private _openings: ICorpJobOpeningData[] = [];
    private _ranks: ICorpRankOption[] = [];

    public flush(): boolean {
        this._corpId = 0;
        this._corpName = '';
        this._isManager = false;
        this._openings = [];
        this._ranks = [];
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        this._corpId = wrapper.readInt();
        this._corpName = wrapper.readString();
        this._isManager = wrapper.readBoolean();

        // openings
        const count = wrapper.readInt();
        this._openings = [];

        for (let i = 0; i < count; i++) {
            const id = wrapper.readInt();
            const title = wrapper.readString();
            const openSlots = wrapper.readInt();
            const isOpen = wrapper.readBoolean();
            const createdByUserId = wrapper.readInt();
            const createdAt = wrapper.readInt();

            this._openings.push({
                id,
                title,
                openSlots,
                isOpen,
                createdByUserId,
                createdAt,
            });
        }

        // ranks (MUST match composer)
        const rankCount = wrapper.readInt();
        this._ranks = [];
        for (let i = 0; i < rankCount; i++) {
            const id = wrapper.readInt();
            const name = wrapper.readString();
            this._ranks.push({ id, name });
        }

        return true;
    }

    public get corpId(): number {
        return this._corpId;
    }
    public get corpName(): string {
        return this._corpName;
    }
    public get isManager(): boolean {
        return this._isManager;
    }
    public get openings(): ICorpJobOpeningData[] {
        return this._openings;
    }
    public get ranks(): ICorpRankOption[] {
        return this._ranks;
    }
}
