// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/OpenFishingModuleParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

/**
 * Parses: [ spotId:int, difficulty:int, durationSeconds:int, spotLabel:string ]
 * from OpenFishingModuleComposer (server -> client).
 */
export class OpenFishingModuleMessageParser implements IMessageParser {
    private _spotId: number = 0;
    private _difficulty: number = 1;
    private _durationSeconds: number = 0;
    private _spotLabel: string = '';

    flush(): boolean {
        this._spotId = 0;
        this._difficulty = 1;
        this._durationSeconds = 0;
        this._spotLabel = '';
        return true;
    }

    parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        this._spotId = wrapper.readInt();
        this._difficulty = wrapper.readInt();
        this._durationSeconds = wrapper.readInt();
        this._spotLabel = wrapper.readString();

        return true;
    }

    get spotId(): number {
        return this._spotId;
    }
    get difficulty(): number {
        return this._difficulty;
    }
    get durationSeconds(): number {
        return this._durationSeconds;
    }
    get spotLabel(): string {
        return this._spotLabel;
    }
}
