// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/OpponentStatsParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export class OpponentStatsParser implements IMessageParser {
    private _userId = 0;
    private _username = '';
    private _figure = '';

    private _health = 0;
    private _maxHealth = 0;

    private _energy = 0;
    private _maxEnergy = 0;

    private _hunger = 0;
    private _maxHunger = 0;

    private _aggression = 0; // 0–100
    private _level = 0;

    public flush(): boolean {
        this._userId = 0;
        this._username = '';
        this._figure = '';
        this._health = this._maxHealth = 0;
        this._energy = this._maxEnergy = 0;
        this._hunger = this._maxHunger = 0;
        this._aggression = 0;
        this._level = 0;
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        this._userId = wrapper.readInt();
        this._username = wrapper.readString();
        this._figure = wrapper.readString();

        this._health = wrapper.readInt();
        this._maxHealth = wrapper.readInt();

        this._energy = wrapper.readInt();
        this._maxEnergy = wrapper.readInt();

        this._hunger = wrapper.readInt();
        this._maxHunger = wrapper.readInt();

        this._aggression = wrapper.readInt();
        this._level = wrapper.readInt();

        return true;
    }

    get userId() {
        return this._userId;
    }
    get username() {
        return this._username;
    }
    get figure() {
        return this._figure;
    }
    get health() {
        return this._health;
    }
    get maxHealth() {
        return this._maxHealth;
    }
    get energy() {
        return this._energy;
    }
    get maxEnergy() {
        return this._maxEnergy;
    }
    get hunger() {
        return this._hunger;
    }
    get maxHunger() {
        return this._maxHunger;
    }
    get aggression() {
        return this._aggression;
    }
    get level() {
        return this._level;
    }
}
