// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/OpponentStatsParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export class OpponentStatsParser implements IMessageParser {
    private _userId = 0;
    private _username = '';
    private _figure = '';

    private _health = 0;
    private _maxHealth = 0;

    private _energy = 0;
    private _maxEnergy = 0;

    private _hunger = 0;
    private _maxHunger = 0;

    private _aggressionMs = 0;

    private _xp = 0;
    private _maxXP = 0;
    private _strength = 0;
    private _level = 0;
    private _points = 0;
    private _stamina = 0;
    private _hungerLevel = 0;
    private _gatheringLevel = 0;
    private _defense = 0;

    private _punchesThrown = 0;
    private _punchesLanded = 0;
    private _damageDealt = 0;
    private _damageReceived = 0;
    private _shiftsWorked = 0;
    private _kills = 0;
    private _deaths = 0;
    private _arrests = 0;

    private _isAggressive = false;
    private _gangName = '';

    private _healthLevel = 0;

    public flush(): boolean {
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        this._userId = wrapper.readInt();
        this._username = wrapper.readString();
        this._figure = wrapper.readString();

        this._health = wrapper.readInt();
        this._maxHealth = wrapper.readInt();

        this._energy = wrapper.readInt();
        this._maxEnergy = wrapper.readInt();

        this._hunger = wrapper.readInt();
        this._maxHunger = wrapper.readInt();

        this._aggressionMs = wrapper.readInt();

        this._xp = wrapper.readInt();
        this._maxXP = wrapper.readInt();
        this._strength = wrapper.readInt();
        this._level = wrapper.readInt();
        this._points = wrapper.readInt();
        this._stamina = wrapper.readInt();
        this._hungerLevel = wrapper.readInt();
        this._gatheringLevel = wrapper.readInt();
        this._defense = wrapper.readInt();

        this._punchesThrown = wrapper.readInt();
        this._punchesLanded = wrapper.readInt();
        this._damageDealt = wrapper.readInt();
        this._damageReceived = wrapper.readInt();
        this._shiftsWorked = wrapper.readInt();
        this._kills = wrapper.readInt();
        this._deaths = wrapper.readInt();
        this._arrests = wrapper.readInt();

        this._isAggressive = wrapper.readBoolean();
        this._gangName = wrapper.readString();

        this._healthLevel = wrapper.readInt();

        return true;
    }

    get userId() {
        return this._userId;
    }
    get username() {
        return this._username;
    }
    get figure() {
        return this._figure;
    }

    get health() {
        return this._health;
    }
    get maxHealth() {
        return this._maxHealth;
    }
    get energy() {
        return this._energy;
    }
    get maxEnergy() {
        return this._maxEnergy;
    }
    get hunger() {
        return this._hunger;
    }
    get maxHunger() {
        return this._maxHunger;
    }

    get aggressionMs() {
        return this._aggressionMs;
    }

    get xp() {
        return this._xp;
    }
    get maxXP() {
        return this._maxXP;
    }
    get strength() {
        return this._strength;
    }
    get level() {
        return this._level;
    }
    get points() {
        return this._points;
    }
    get stamina() {
        return this._stamina;
    }
    get hungerLevel() {
        return this._hungerLevel;
    }
    get gatheringLevel() {
        return this._gatheringLevel;
    }
    get defense() {
        return this._defense;
    }

    get punchesThrown() {
        return this._punchesThrown;
    }
    get punchesLanded() {
        return this._punchesLanded;
    }
    get damageDealt() {
        return this._damageDealt;
    }
    get damageReceived() {
        return this._damageReceived;
    }
    get shiftsWorked() {
        return this._shiftsWorked;
    }
    get kills() {
        return this._kills;
    }
    get deaths() {
        return this._deaths;
    }
    get arrests() {
        return this._arrests;
    }

    get isAggressive() {
        return this._isAggressive;
    }
    get gangName() {
        return this._gangName;
    }

    get healthLevel() {
        return this._healthLevel;
    }
}
