// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/PaletteHexesParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export interface PaletteHexesData {
    part: string; // e.g. "cc"
    paletteId: number; // e.g. 2
    hexes: string[]; // e.g. ["FFFFFF","000000", ...]
}

export class PaletteHexesParser implements IMessageParser {
    private _data: PaletteHexesData | null = null;

    public flush(): boolean {
        this._data = null;
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        // order must match the server composer:
        // string part, int paletteId, int count, count * string hex
        const part = wrapper.readString();
        const paletteId = wrapper.readInt();

        const count = wrapper.readInt();
        const hexes: string[] = new Array(count);

        for (let i = 0; i < count; i++) {
            // Server sends hex without '#', but this parser accepts either
            let raw = wrapper.readString();
            if (!raw) raw = '';
            hexes[i] = raw.replace('#', '').toUpperCase();
        }

        this._data = { part, paletteId, hexes };
        return true;
    }

    public get data(): PaletteHexesData | null {
        return this._data;
    }

    // Convenience getters (optional)
    public get part(): string | null {
        return this._data?.part ?? null;
    }

    public get paletteId(): number {
        return this._data?.paletteId ?? -1;
    }

    public get hexes(): string[] {
        return this._data?.hexes ?? [];
    }
}
