// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/PoliceCallParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export interface PoliceCallData {
    username: string;
    figure: string;
    message: string;
    virtualRoomId: number;
    virtualRoomName: string;
}

export class PoliceCallParser implements IMessageParser {
    private _data: PoliceCallData = {
        username: '',
        figure: '',
        message: '',
        virtualRoomId: 0,
        virtualRoomName: '',
    };

    public flush(): boolean {
        this._data = {
            username: '',
            figure: '',
            message: '',
            virtualRoomId: 0,
            virtualRoomName: '',
        };
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        // MUST match server composer field order:
        // username:string, figure:string, message:string, vId:int, vName:string
        const username = wrapper.readString();
        const figure = wrapper.readString();
        const message = wrapper.readString();
        const virtualRoomId = wrapper.readInt();
        const virtualRoomName = wrapper.readString();

        this._data = {
            username,
            figure,
            message,
            virtualRoomId,
            virtualRoomName,
        };
        return true;
    }

    public get data(): PoliceCallData {
        return this._data;
    }
}
