// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/ProfileAchievementsParser.ts
import { IMessageDataWrapper, IMessageEvent } from '@nitrots/nitro-renderer';

export interface AchievementRow {
    id: number;
    key: string;
    name: string;
    current: number;
    target: number;
    xpReward: number;
    badgeUrl: string;
    description?: string;

    // 🔹 NEW: grouping + ordering
    groupKey: string; // 'kills', 'deaths', 'punches', etc.
    groupOrder: number; // group sort order (10, 20, 30...)
    levelOrder: number; // level within group (1..10)
}

export class ProfileAchievementsParser {
    private _achievements: AchievementRow[] = [];

    public flush(): boolean {
        this._achievements = [];
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        this._achievements = [];

        const count = wrapper.readInt();

        for (let i = 0; i < count; i++) {
            const id = wrapper.readInt();
            const key = wrapper.readString();
            const name = wrapper.readString();
            const current = wrapper.readInt();
            const target = wrapper.readInt();
            const xpReward = wrapper.readInt();
            const badgeUrl = wrapper.readString();
            const description = wrapper.readString();

            // 🔹 NEW: read group & order
            const groupKey = wrapper.readString();
            const groupOrder = wrapper.readInt();
            const levelOrder = wrapper.readInt();

            this._achievements.push({
                id,
                key,
                name,
                current,
                target,
                xpReward,
                badgeUrl,
                description: description || undefined,
                groupKey: groupKey || 'other',
                groupOrder,
                levelOrder,
            });
        }

        // fire a DOM event so MyProfileView or a store can pick it up
        window.dispatchEvent(
            new CustomEvent('profile_achievements_update', {
                detail: { achievements: this._achievements },
            })
        );

        return true;
    }

    public get achievements(): AchievementRow[] {
        return this._achievements;
    }
}
