// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/SlotMachineStateParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export type SlotStateStr = 'IDLE' | 'SPINNING' | 'ENDED';

export class SlotMachineStateParser implements IMessageParser {
    private _state: SlotStateStr = 'IDLE';
    private _reels = 3;
    private _stops: number[] = [];
    private _win = 0;
    private _bet = 0;
    private _messages: { text: string; level: 'info' | 'warn' | 'error' }[] =
        [];

    public flush(): boolean {
        this._state = 'IDLE';
        this._reels = 3;
        this._stops = [];
        this._win = 0;
        this._bet = 0;
        this._messages = [];
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        // server sends: state (STRING), reels (INT), stops (INT[]), win (INT), bet (INT), messages ([text,level] pairs)
        this._state = (
            wrapper.readString() || 'IDLE'
        ).toUpperCase() as SlotStateStr;
        this._reels = wrapper.readInt();

        const stopsLen = wrapper.readInt();
        this._stops = new Array(stopsLen);
        for (let i = 0; i < stopsLen; i++) this._stops[i] = wrapper.readInt();

        this._win = wrapper.readInt();
        this._bet = wrapper.readInt();

        const msgLen = wrapper.readInt();
        this._messages = new Array(msgLen);
        for (let i = 0; i < msgLen; i++) {
            const text = wrapper.readString();
            const level = wrapper.readString() as 'info' | 'warn' | 'error';
            this._messages[i] = { text, level };
        }

        return true;
    }

    public get state(): SlotStateStr {
        return this._state;
    }
    public get reels(): number {
        return this._reels;
    }
    public get stops(): number[] {
        return this._stops;
    }
    public get win(): number {
        return this._win;
    }
    public get bet(): number {
        return this._bet;
    }
    public get messages() {
        return this._messages;
    }
}
