// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/TokenStoreParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export interface TokenShopItem {
    id: number;
    key: string; // e.g. "DOUBLE_XP"
    title: string; // e.g. "Double XP (12h)"
    durationHours: number; // 12 / 24
    price: number; // 300 / 600
    imageKey: string; // optional path/key
    owned: boolean; // (future use)
    activeUntilTs: number; // epoch millis or 0
}

export class TokensStoreParser implements IMessageParser {
    private _diamonds: number = 0;
    private _items: TokenShopItem[] = [];

    public flush(): boolean {
        this._diamonds = 0;
        this._items = [];
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        this._diamonds = wrapper.readInt();

        const count = wrapper.readInt();
        this._items = new Array<TokenShopItem>(count);

        for (let i = 0; i < count; i++) {
            const id = wrapper.readInt();
            const key = wrapper.readString();
            const title = wrapper.readString();
            const durationHours = wrapper.readInt();
            const price = wrapper.readInt();
            const imageKey = wrapper.readString();
            const owned = wrapper.readBoolean();
            const activeUntilTs = wrapper.readInt();

            this._items[i] = {
                id,
                key,
                title,
                durationHours,
                price,
                imageKey,
                owned,
                activeUntilTs,
            };
        }

        return true;
    }

    public get diamonds(): number {
        return this._diamonds;
    }
    public get items(): TokenShopItem[] {
        return this._items;
    }
}
