// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/UserLiveFeedParser.ts
import {
    IMessageDataWrapper,
    IMessageParser,
} from '@nitrots/nitro-renderer/src/core';

export interface UserLiveFeedData {
    username: string;
    figure: string;
    message: string;
    type: string;
    persistent: boolean;
}

export class UserLiveFeedParser implements IMessageParser {
    private _data: UserLiveFeedData = {
        username: '',
        figure: '',
        message: '',
        type: '',
        persistent: false
    };

    public flush(): boolean {
        this._data = {
            username: '',
            figure: '',
            message: '',
            type: '',
            persistent: false
        };
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        try {
            const username = wrapper.readString();
            const figure = wrapper.readString();
            const message = wrapper.readString();
            const type = wrapper.readString();
            const persistent = wrapper.readBoolean();

            this._data = { username, figure, message, type, persistent };

            return true;
        } catch (e) {
            console.error('[❌] Failed to parse LiveFeed packet:', e);
            return false;
        }
    }

    public get data(): UserLiveFeedData {
        return this._data;
    }
}
