// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/WantedListParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export interface WantedUserData {
    userId: number;
    username: string;
    figure: string;
    wantedLevel: number;
    charges: string; // comma-separated (raw)
    chargesList: string[]; // parsed array of charges
}

export class WantedListParser implements IMessageParser {
    private _data: WantedUserData[] = [];

    public flush(): boolean {
        this._data = [];
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        const count = wrapper.readInt();
        const list: WantedUserData[] = [];

        for (let i = 0; i < count; i++) {
            const userId = wrapper.readInt();
            const username = wrapper.readString();
            const figure = wrapper.readString();
            const wantedLevel = wrapper.readInt();
            const charges = wrapper.readString();

            const chargesList = charges
                ? charges
                      .split(',')
                      .map((s) => s.trim())
                      .filter(Boolean)
                : [];

            list.push({
                userId,
                username,
                figure,
                wantedLevel,
                charges,
                chargesList,
            });
        }

        this._data = list;

        // 🔥 Broadcast event so UI updates immediately
        window.dispatchEvent(
            new CustomEvent('wanted_list_update', { detail: this._data })
        );

        return true;
    }

    public getWantedUsers(): WantedUserData[] {
        return this._data;
    }
}
