// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/WantedListParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export interface WantedUserData {
    userId: number;
    username: string;
    figure: string;
    wantedLevel: number;
    charges: string;
}

export class WantedListParser implements IMessageParser {
    private _data: WantedUserData[] = [];

    public flush(): boolean {
        this._data = [];
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        const count = wrapper.readInt();

        for (let i = 0; i < count; i++) {
            this._data.push({
                userId: wrapper.readInt(),
                username: wrapper.readString(),
                figure: wrapper.readString(),
                wantedLevel: wrapper.readInt(),
                charges: wrapper.readString(),
            });
        }
        // 🔥 Dispatch event so React can listen
        window.dispatchEvent(
            new CustomEvent('wanted_list_update', { detail: this._data })
        );

        return true;
    }

    public getWantedUsers(): WantedUserData[] {
        return this._data;
    }
}
