// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/WheelModuleParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export class WheelModuleParser implements IMessageParser {
    private _hasInfo = false;
    private _faces: string[] = [];
    private _payouts: Record<string, number> = {};
    private _minBet = 1;
    private _maxBet = 5000;
    private _spinMs = 2500;

    public flush(): boolean {
        this._hasInfo = false;
        this._faces = [];
        this._payouts = {};
        this._minBet = 1;
        this._maxBet = 5000;
        this._spinMs = 2500;
        return true;
    }

    public parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;

        this._hasInfo = wrapper.readBoolean();

        const nFaces = wrapper.readInt();
        this._faces = new Array(nFaces);
        for (let i = 0; i < nFaces; i++) this._faces[i] = wrapper.readString();

        const nPay = wrapper.readInt();
        this._payouts = {};
        for (let i = 0; i < nPay; i++) {
            const key = wrapper.readString();
            const mul = wrapper.readInt();
            this._payouts[key] = mul;
        }

        this._minBet = wrapper.readInt();
        this._maxBet = wrapper.readInt();
        this._spinMs = wrapper.readInt();

        return true;
    }

    public get hasInfo() {
        return this._hasInfo;
    }
    public get faces() {
        return this._faces;
    }
    public get payouts() {
        return this._payouts;
    }
    public get minBet() {
        return this._minBet;
    }
    public get maxBet() {
        return this._maxBet;
    }
    public get spinMs() {
        return this._spinMs;
    }
}
