// @nitrots/nitro-renderer/src/nitro/communication/messages/parser/WheelStateParser.ts
import { IMessageDataWrapper, IMessageParser } from '@nitrots/nitro-renderer';

export class WheelStateParser implements IMessageParser {
    private _state: 'SPINNING' | 'IDLE' = 'IDLE';
    private _landedIndex = -1;
    private _landedFace = '';
    private _win = 0;
    private _betAmount = 0;
    private _betFace = '';
    private _faces: string[] = [];

    flush(): boolean {
        this._state = 'IDLE';
        this._landedIndex = -1;
        this._landedFace = '';
        this._win = 0;
        this._betAmount = 0;
        this._betFace = '';
        this._faces = [];
        return true;
    }

    parse(wrapper: IMessageDataWrapper): boolean {
        if (!wrapper) return false;
        this._state = (wrapper.readString() || 'IDLE') as any;
        this._landedIndex = wrapper.readInt();
        this._landedFace = wrapper.readString();
        this._win = wrapper.readInt();
        this._betAmount = wrapper.readInt();
        this._betFace = wrapper.readString();
        const n = wrapper.readInt();
        this._faces = [];
        for (let i = 0; i < n; i++) this._faces.push(wrapper.readString());
        return true;
    }

    get state() {
        return this._state;
    }
    get landedIndex() {
        return this._landedIndex;
    }
    get landedFace() {
        return this._landedFace;
    }
    get win() {
        return this._win;
    }
    get betAmount() {
        return this._betAmount;
    }
    get betFace() {
        return this._betFace;
    }
    get faces() {
        return this._faces;
    }
}
